// sensd/es/dropdown/dropdown-button.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React, { useRef, useState, useEffect } from 'react';
import classNames from 'classnames';
import MoreOutlined from "@sensd/icons/es/icons/MoreOutlined";
import LoadingOutlined from "@sensd/icons/es/icons/LoadingOutlined";
import DownFilled from "@sensd/icons/es/icons/DownFilled";
import UpFilled from "@sensd/icons/es/icons/UpFilled";
import Button from '../button';
import { ConfigContext } from '../config-provider';
import Dropdown from './dropdown';

var isArrowButtonType = function isArrowButtonType(type) {
  return type === 'link';
};

var isOnlyHoverTrigger = function isOnlyHoverTrigger(trigger) {
  if (!trigger) {
    return true;
  }

  if (Array.isArray(trigger) && trigger.length === 1 && trigger[0] === 'hover') {
    return true;
  }

  return false;
};

var DropdownButton = function DropdownButton(props) {
  var _React$useContext = React.useContext(ConfigContext),
      getContextPopupContainer = _React$useContext.getPopupContainer,
      getPrefixCls = _React$useContext.getPrefixCls,
      direction = _React$useContext.direction;

  var customizePrefixCls = props.prefixCls,
      _props$type = props.type,
      type = _props$type === void 0 ? 'default' : _props$type,
      disabled = props.disabled,
      loading = props.loading,
      onClick = props.onClick,
      htmlType = props.htmlType,
      children = props.children,
      className = props.className,
      overlay = props.overlay,
      trigger = props.trigger,
      align = props.align,
      visible = props.visible,
      onVisibleChange = props.onVisibleChange,
      placement = props.placement,
      getPopupContainer = props.getPopupContainer,
      href = props.href,
      icon = props.icon,
      title = props.title,
      _props$buttonsRender = props.buttonsRender,
      buttonsRender = _props$buttonsRender === void 0 ? function (buttons) {
    return buttons;
  } : _props$buttonsRender,
      mouseEnterDelay = props.mouseEnterDelay,
      mouseLeaveDelay = props.mouseLeaveDelay,
      overlayClassName = props.overlayClassName,
      overlayStyle = props.overlayStyle,
      destroyPopupOnHide = props.destroyPopupOnHide,
      _props$noWave = props.noWave,
      noWave = _props$noWave === void 0 ? false : _props$noWave,
      restProps = __rest(props, ["prefixCls", "type", "disabled", "loading", "onClick", "htmlType", "children", "className", "overlay", "trigger", "align", "visible", "onVisibleChange", "placement", "getPopupContainer", "href", "icon", "title", "buttonsRender", "mouseEnterDelay", "mouseLeaveDelay", "overlayClassName", "overlayStyle", "destroyPopupOnHide", "noWave"]);

  var prefixCls = getPrefixCls('dropdown-button', customizePrefixCls);
  var overlayPrefixCls = getPrefixCls('dropdown', customizePrefixCls);
  var innerDisabled = disabled || loading;
  var triggerRef = useRef(null);
  var listenerRef = useRef(null);

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1];

  var handleVisibleChange = function handleVisibleChange(value) {
    setOpen(value);
    onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(value);
  };

  var getPopupDomNode = function getPopupDomNode() {
    var _a, _b;

    return (_b = (_a = triggerRef === null || triggerRef === void 0 ? void 0 : triggerRef.current) === null || _a === void 0 ? void 0 : _a.getPopupDomNode) === null || _b === void 0 ? void 0 : _b.call(_a);
  };

  var clearListener = function clearListener() {
    var popupDomNode = getPopupDomNode();

    if (listenerRef.current && popupDomNode) {
      popupDomNode.removeEventListener(listenerRef.current);
      listenerRef.current = null;
    }
  };

  var mergedTrigger = trigger || (isArrowButtonType(type) ? ['click'] : undefined);
  var onlyHoverTrigger = isOnlyHoverTrigger(mergedTrigger);
  var dropdownProps = {
    align: align,
    overlay: overlay,
    disabled: innerDisabled,
    trigger: disabled ? [] : mergedTrigger,
    onVisibleChange: handleVisibleChange,
    getPopupContainer: getPopupContainer || getContextPopupContainer,
    mouseEnterDelay: mouseEnterDelay,
    mouseLeaveDelay: mouseLeaveDelay,
    overlayClassName: classNames(overlayClassName, _defineProperty({}, "".concat(overlayPrefixCls, "-arrow-trigger"), isArrowButtonType(type))),
    overlayStyle: overlayStyle,
    destroyPopupOnHide: destroyPopupOnHide
  };

  if ('visible' in props) {
    dropdownProps.visible = visible;
  }

  if ('placement' in props) {
    dropdownProps.placement = placement;
  } else {
    dropdownProps.placement = direction === 'rtl' ? 'bottomLeft' : 'bottomRight';
  } // fix: 监听点击 menu 收起菜单


  useEffect(function () {
    if (!open) {
      clearListener();
      return;
    }

    var popupDomNode = getPopupDomNode();

    if (!popupDomNode) {
      return;
    }

    clearListener();
    listenerRef.current = popupDomNode.addEventListener('click', function () {
      setTimeout(function () {
        var _a, _b;

        var popupVisible = (_b = (_a = triggerRef.current) === null || _a === void 0 ? void 0 : _a.state) === null || _b === void 0 ? void 0 : _b.popupVisible;

        if (popupVisible !== open) {
          setOpen(popupVisible);
        }
      });
    });
  }, [open]);
  useEffect(function () {
    return function () {
      clearListener();
    };
  }, []);
  useEffect(function () {
    if (innerDisabled && triggerRef.current) {
      // 禁用的时候默认收起弹窗
      // 依赖 trigger 组件实例方法
      triggerRef.current.setPopupVisible(false);
      setOpen(false);
    }
  }, [innerDisabled]);

  var renderSuffixIcon = function renderSuffixIcon() {
    if (loading) return /*#__PURE__*/React.createElement(LoadingOutlined, null);
    if (icon) return icon;

    if (isArrowButtonType(type)) {
      if (open) return /*#__PURE__*/React.createElement(UpFilled, null);
      return /*#__PURE__*/React.createElement(DownFilled, null);
    }

    return /*#__PURE__*/React.createElement(MoreOutlined, null);
  };

  var cls = classNames(prefixCls, className, _defineProperty({}, "".concat(prefixCls, "-only-hover"), onlyHoverTrigger));
  var button = /*#__PURE__*/React.createElement(Button, _extends({}, restProps, {
    className: cls,
    type: type,
    onClick: onClick,
    htmlType: htmlType,
    href: href,
    title: title,
    noWave: noWave || onlyHoverTrigger
  }), children, renderSuffixIcon());

  var _buttonsRender = buttonsRender([button, []]),
      _buttonsRender2 = _slicedToArray(_buttonsRender, 1),
      buttonToRender = _buttonsRender2[0];

  return /*#__PURE__*/React.createElement(Dropdown, _extends({
    ref: triggerRef
  }, dropdownProps), buttonToRender);
};

DropdownButton.__SENS_BUTTON = true;
export default DropdownButton;
