// sensd/es/upload/UploadList/ListItem.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import * as React from 'react';
import CSSMotion from 'rc-motion';
import classNames from 'classnames';
import VisibleOutlined from "@sensd/icons/es/icons/VisibleOutlined";
import DeleteOutlined from "@sensd/icons/es/icons/DeleteOutlined";
import ErrorFilled from "@sensd/icons/es/icons/ErrorFilled";
import DownloadOutlined from "@sensd/icons/es/icons/DownloadOutlined";
import Typography from '../../typography';
import ItemProgress from './ItemProgress';
import Tooltip from '../../tooltip';
import { ConfigContext } from '../../config-provider';
import { getSuitableFileUnit, formatSizeToSuitableUnit } from '../utils';
import { ErrorType } from '../interface';

var EllipsisMiddle = function EllipsisMiddle(_ref) {
  var suffixCount = _ref.suffixCount,
      children = _ref.children;
  var dotIndex = children.lastIndexOf('.') || children.length;
  var extLength = children.slice(dotIndex).length;
  var start = children.slice(0, children.length - suffixCount - extLength).trim();
  var suffix = children.slice(-suffixCount - extLength).trim();
  return /*#__PURE__*/React.createElement(Typography.Text, {
    style: {
      maxWidth: '100%'
    },
    ellipsis: {
      suffix: suffix,
      tooltip: children
    }
  }, start);
};

var ListItem = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
  var _classNames3;

  var prefixCls = _ref2.prefixCls,
      className = _ref2.className,
      style = _ref2.style,
      locale = _ref2.locale,
      type = _ref2.type,
      listType = _ref2.listType,
      file = _ref2.file,
      items = _ref2.items,
      progressProps = _ref2.progress,
      iconRender = _ref2.iconRender,
      actionIconRender = _ref2.actionIconRender,
      itemRender = _ref2.itemRender,
      isImgUrl = _ref2.isImgUrl,
      showPreviewIcon = _ref2.showPreviewIcon,
      showRemoveIcon = _ref2.showRemoveIcon,
      showCancelIcon = _ref2.showCancelIcon,
      showDownloadIcon = _ref2.showDownloadIcon,
      customPreviewIcon = _ref2.previewIcon,
      customRemoveIcon = _ref2.removeIcon,
      customCancelIcon = _ref2.cancelIcon,
      customDownloadIcon = _ref2.downloadIcon,
      isSinglePicture = _ref2.isSinglePicture,
      showFileSize = _ref2.showFileSize,
      onPreview = _ref2.onPreview,
      onDownload = _ref2.onDownload,
      onClose = _ref2.onClose;

  var _a, _b; // Delay to show the progress bar


  var _React$useState = React.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      showProgress = _React$useState2[0],
      setShowProgress = _React$useState2[1]; // when listType is 'picture-card' and non drag single picture, progressType is 'picture'


  var progressType = listType === 'picture-card' && !(type === 'drag' && isSinglePicture) ? 'picture' : 'text';
  var progressRafRef = React.useRef();
  React.useEffect(function () {
    progressRafRef.current = setTimeout(function () {
      setShowProgress(true);
    }, 300);
    return function () {
      window.clearTimeout(progressRafRef.current);
    };
  }, []); // This is used for legacy span make scrollHeight the wrong value.
  // We will force these to be `display: block` with non `picture-card`

  var spanClassName = "".concat(prefixCls, "-span");
  var iconNode = iconRender(file);
  var icon = /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-text-icon")
  }, iconNode);

  if (listType === 'picture' || listType === 'picture-card') {
    if (listType === 'picture-card' && (file.status === 'uploading' || file.status === 'error')) {
      icon = null;
    } else if (!file.thumbUrl && !file.url) {
      var _classNames;

      var uploadingClassName = classNames((_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-list-item-thumbnail"), true), _defineProperty(_classNames, "".concat(prefixCls, "-list-item-file"), true), _classNames));
      icon = /*#__PURE__*/React.createElement("div", {
        className: uploadingClassName
      }, iconNode);
    } else {
      var _classNames2;

      var thumbnail = (isImgUrl === null || isImgUrl === void 0 ? void 0 : isImgUrl(file)) ? /*#__PURE__*/React.createElement("img", {
        src: file.thumbUrl || file.url,
        alt: file.name,
        className: "".concat(prefixCls, "-list-item-image")
      }) : iconNode;
      var aClassName = classNames((_classNames2 = {}, _defineProperty(_classNames2, "".concat(prefixCls, "-list-item-thumbnail"), true), _defineProperty(_classNames2, "".concat(prefixCls, "-list-item-file"), isImgUrl && !isImgUrl(file)), _classNames2));
      icon = /*#__PURE__*/React.createElement("a", {
        className: aClassName,
        onClick: function onClick(e) {
          return onPreview(file, e);
        },
        href: file.url || file.thumbUrl,
        target: "_blank",
        rel: "noopener noreferrer"
      }, thumbnail);
    }
  }

  var infoUploadingClass = classNames((_classNames3 = {}, _defineProperty(_classNames3, "".concat(prefixCls, "-list-item"), true), _defineProperty(_classNames3, "".concat(prefixCls, "-list-item-").concat(file.status), true), _defineProperty(_classNames3, "".concat(prefixCls, "-list-item-type-").concat(listType), true), _defineProperty(_classNames3, "".concat(prefixCls, "-list-item-is-file"), !file.thumbUrl && !file.url || isImgUrl && !isImgUrl(file)), _classNames3));
  var linkProps = typeof file.linkProps === 'string' ? JSON.parse(file.linkProps) : file.linkProps;
  var removeIcon = showRemoveIcon ? actionIconRender((typeof customRemoveIcon === 'function' ? customRemoveIcon(file) : customRemoveIcon) || (listType !== 'picture-card' || progressType === 'text' && file.status === 'error' ? locale.remove : /*#__PURE__*/React.createElement(Tooltip, {
    title: locale.remove
  }, /*#__PURE__*/React.createElement(DeleteOutlined, null))), function () {
    return onClose(file);
  }, prefixCls, locale.removeFile) : null;
  var cancelIcon = showCancelIcon ? actionIconRender((typeof customCancelIcon === 'function' ? customCancelIcon(file) : customCancelIcon) || locale.cancel, function () {
    return onClose(file);
  }, prefixCls, locale.cancelUpload) : null;
  var downloadIcon = showDownloadIcon && file.status === 'done' ? actionIconRender((typeof customDownloadIcon === 'function' ? customDownloadIcon(file) : customDownloadIcon) || (listType === 'picture-card' ? /*#__PURE__*/React.createElement(Tooltip, {
    title: locale.download
  }, /*#__PURE__*/React.createElement(DownloadOutlined, null)) : locale.download), function () {
    return onDownload(file);
  }, prefixCls, locale.downloadFile) : null;
  var downloadOrDelete = listType !== 'picture-card' && file.status === 'done' && /*#__PURE__*/React.createElement("span", {
    key: "download-delete",
    className: classNames("".concat(prefixCls, "-list-item-card-actions"), {
      picture: listType === 'picture'
    })
  }, downloadIcon, removeIcon);
  var fileSizeUnit = getSuitableFileUnit(file.size || 0);
  var sizeSpan = file.size && file.status === 'done' && showFileSize ? /*#__PURE__*/React.createElement("span", {
    key: "size",
    className: "".concat(prefixCls, "-list-item-size")
  }, formatSizeToSuitableUnit(file.size, fileSizeUnit)) : null;
  var nameText = listType === 'picture-card' ? file.name : /*#__PURE__*/React.createElement(EllipsisMiddle, {
    suffixCount: 2
  }, file.name);
  var listItemNameClass = classNames("".concat(prefixCls, "-list-item-name"));
  var preview = file.url ? [/*#__PURE__*/React.createElement("a", _extends({
    key: "view",
    target: "_blank",
    rel: "noopener noreferrer",
    className: listItemNameClass
  }, linkProps, {
    href: file.url,
    onClick: function onClick(e) {
      return onPreview(file, e);
    }
  }), file.name ? nameText : null), sizeSpan, downloadOrDelete] : [/*#__PURE__*/React.createElement("span", {
    key: "view",
    className: listItemNameClass,
    onClick: function onClick(e) {
      return onPreview(file, e);
    }
  }, file.name ? nameText : null), sizeSpan, downloadOrDelete];
  var previewStyle = {
    pointerEvents: 'none',
    opacity: 0.5
  };
  var previewIcon = showPreviewIcon ? /*#__PURE__*/React.createElement("a", {
    href: file.url || file.thumbUrl,
    target: "_blank",
    rel: "noopener noreferrer",
    style: file.url || file.thumbUrl ? undefined : previewStyle,
    onClick: function onClick(e) {
      return onPreview(file, e);
    },
    title: locale.previewFile
  }, typeof customPreviewIcon === 'function' ? customPreviewIcon(file) : customPreviewIcon || /*#__PURE__*/React.createElement(Tooltip, {
    title: locale.preview
  }, /*#__PURE__*/React.createElement(VisibleOutlined, null))) : null;
  var actions = listType === 'picture-card' && file.status !== 'uploading' && /*#__PURE__*/React.createElement("span", {
    className: "".concat(prefixCls, "-list-item-actions")
  }, file.status !== 'error' && previewIcon, file.status === 'done' && downloadIcon, removeIcon);
  var overFileSizeMsg = listType === 'picture-card' && (isImgUrl === null || isImgUrl === void 0 ? void 0 : isImgUrl(file)) ? locale.pictureOverSize : locale.fileOverSize;
  var errorMsg = ((_a = file.error) === null || _a === void 0 ? void 0 : _a.statusText) || ((_b = file.error) === null || _b === void 0 ? void 0 : _b.message) || (file.errorType === ErrorType.OverMaxSize ? overFileSizeMsg : locale.uploadError);
  var errorDom = null;

  if (file.status === 'error') {
    errorDom = /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-list-item-status")
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        maxWidth: '100%'
      }
    }, /*#__PURE__*/React.createElement(ErrorFilled, null), /*#__PURE__*/React.createElement("span", {
      className: "".concat(prefixCls, "-list-item-error-msg")
    }, errorMsg)), progressType !== 'picture' && removeIcon);
  }

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var rootPrefixCls = getPrefixCls();
  var dom = /*#__PURE__*/React.createElement("div", {
    className: infoUploadingClass
  }, icon, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-list-item-info")
  }, /*#__PURE__*/React.createElement("span", {
    className: spanClassName
  }, preview), file.status === 'error' && progressType === 'picture' && errorDom, showProgress && /*#__PURE__*/React.createElement(CSSMotion, {
    motionName: "".concat(rootPrefixCls, "-fade"),
    visible: file.status === 'uploading' || file.status === 'error' && progressType === 'text',
    motionDeadline: 2000
  }, function (_ref3) {
    var motionClassName = _ref3.className;
    return /*#__PURE__*/React.createElement(ItemProgress, {
      type: progressType,
      prefixCls: prefixCls,
      locale: locale,
      file: file,
      cancelIcon: cancelIcon,
      errorDom: errorDom,
      progress: progressProps,
      className: motionClassName
    });
  })), actions);
  var listContainerNameClass = classNames("".concat(prefixCls, "-list-").concat(listType, "-container"), className);
  return /*#__PURE__*/React.createElement("div", {
    className: listContainerNameClass,
    style: style,
    ref: ref
  }, itemRender ? itemRender(dom, file, items, {
    download: onDownload.bind(null, file),
    preview: onPreview.bind(null, file),
    remove: onClose.bind(null, file)
  }) : dom);
});
export default ListItem;
