// sensd/lib/result/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CheckCircleFilled = _interopRequireDefault(require("@sensd/icons/CheckCircleFilled"));

var _CloseCircleFilled = _interopRequireDefault(require("@sensd/icons/CloseCircleFilled"));

var _ExclamationCircleFilled = _interopRequireDefault(require("@sensd/icons/ExclamationCircleFilled"));

var _WarningFilled = _interopRequireDefault(require("@sensd/icons/WarningFilled"));

var _spin = _interopRequireDefault(require("../spin"));

var _NoFound = _interopRequireDefault(require("./NoFound"));

var _ServerError = _interopRequireDefault(require("./ServerError"));

var _Unauthorized = _interopRequireDefault(require("./Unauthorized"));

var _NoData = _interopRequireDefault(require("./NoData"));

var _NetworkError = _interopRequireDefault(require("./NetworkError"));

var _NoResult = _interopRequireDefault(require("./NoResult"));

var _LoadingFail = _interopRequireDefault(require("./LoadingFail"));

var _configProvider = require("../config-provider");

var _devWarning = _interopRequireDefault(require("../_util/devWarning"));

var _type = require("../_util/type");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var IconMap = {
  success: _CheckCircleFilled["default"],
  error: _CloseCircleFilled["default"],
  info: _ExclamationCircleFilled["default"],
  warning: _WarningFilled["default"]
};
var ExceptionMap = {
  '404': _NoFound["default"],
  '500': _ServerError["default"],
  '403': _Unauthorized["default"]
};
var types = (0, _type.tuple)('page', 'component');
var SizeTypes = (0, _type.tuple)('large', 'small'); // ExceptionImageMap keys

var ExceptionStatus = Object.keys(ExceptionMap);

var getIconSize = function getIconSize() {
  var type = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'page';
  var size = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'large';

  switch (type) {
    case 'page':
      return size === 'large' ? '266px' : '192px';

    case 'component':
      return size === 'large' ? '100px' : '50px';

    default:
      return '266px';
  }
};
/**
 * Render icon if ExceptionStatus includes ,render svg image else render iconNode
 *
 * @param prefixCls
 * @param {status, icon}
 */


var renderIcon = function renderIcon(prefixCls, _ref) {
  var status = _ref.status,
      icon = _ref.icon,
      size = _ref.size,
      type = _ref.type,
      IconModule = _ref.IconModule;

  var _a;

  var className = (0, _classnames["default"])("".concat(prefixCls, "-icon"));
  (0, _devWarning["default"])(!(typeof icon === 'string' && icon.length > 2), 'Result', "`icon` is using ReactNode instead of string naming in v4. Please check `".concat(icon, "` at https://ant.design/components/icon"));

  if (ExceptionStatus.includes("".concat(status))) {
    var SVGComponent = null;

    if (IconModule) {
      SVGComponent = IconModule;
    }

    return /*#__PURE__*/React.createElement("div", {
      className: "".concat(className, " ").concat(prefixCls, "-image ").concat(prefixCls, "-image-").concat(type)
    }, SVGComponent ? /*#__PURE__*/React.createElement(SVGComponent, {
      type: type,
      size: getIconSize(type, size)
    }) : /*#__PURE__*/React.createElement(_spin["default"], null));
  }

  var iconNode = /*#__PURE__*/React.createElement((_a = IconMap[status]) !== null && _a !== void 0 ? _a : IconMap.success, {
    style: {
      fontSize: getIconSize(type, size)
    }
  });
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(className, " ").concat(prefixCls, "-image-").concat(type)
  }, icon || iconNode);
};

var renderTitle = function renderTitle(prefixCls, _ref2) {
  var title = _ref2.title,
      type = _ref2.type,
      size = _ref2.size;
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-title ").concat(prefixCls, "-title-").concat(type, " ").concat(prefixCls, "-title-").concat(type, "-").concat(size)
  }, title);
};

var renderSubTitle = function renderSubTitle(prefixCls, _ref3) {
  var subTitle = _ref3.subTitle,
      type = _ref3.type,
      size = _ref3.size;
  if (!subTitle) return null;
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-subtitle ").concat(prefixCls, "-subtitle-").concat(type, " ").concat(prefixCls, "-subtitle-").concat(type, "-").concat(size)
  }, subTitle);
};

var renderExtra = function renderExtra(prefixCls, _ref4) {
  var extra = _ref4.extra;
  return extra && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-extra")
  }, extra);
};

var Result = function Result(_ref5) {
  var customizePrefixCls = _ref5.prefixCls,
      customizeClassName = _ref5.className,
      subTitle = _ref5.subTitle,
      title = _ref5.title,
      style = _ref5.style,
      children = _ref5.children,
      _ref5$status = _ref5.status,
      status = _ref5$status === void 0 ? 'info' : _ref5$status,
      icon = _ref5.icon,
      extra = _ref5.extra,
      _ref5$type = _ref5.type,
      type = _ref5$type === void 0 ? 'page' : _ref5$type,
      _ref5$size = _ref5.size,
      size = _ref5$size === void 0 ? 'large' : _ref5$size;

  var _React$useContext = React.useContext(_configProvider.ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls,
      direction = _React$useContext.direction;

  var prefixCls = getPrefixCls('result', customizePrefixCls);
  var className = (0, _classnames["default"])(prefixCls, "".concat(prefixCls, "-").concat(type), "".concat(prefixCls, "-").concat(type, "-").concat(size), "".concat(prefixCls, "-").concat(status), customizeClassName, (0, _defineProperty2["default"])({}, "".concat(prefixCls, "-rtl"), direction === 'rtl'));
  var IconModule = ExceptionMap[status];
  return /*#__PURE__*/React.createElement("div", {
    className: className,
    style: style
  }, renderIcon(prefixCls, {
    status: status,
    icon: icon,
    size: size,
    type: type,
    IconModule: IconModule
  }), renderTitle(prefixCls, {
    title: title,
    type: type,
    size: size
  }), renderSubTitle(prefixCls, {
    subTitle: subTitle,
    type: type,
    size: size,
    extra: extra
  }), renderExtra(prefixCls, {
    extra: extra
  }), children && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-content")
  }, children));
};

var generatePresetComponent = function generatePresetComponent(Component, defaultType) {
  return function (props) {
    var _props$type = props.type,
        type = _props$type === void 0 ? 'page' : _props$type,
        _props$size = props.size,
        size = _props$size === void 0 ? 'large' : _props$size;
    var resultType = defaultType !== null && defaultType !== void 0 ? defaultType : type;
    var icon = Component ? /*#__PURE__*/React.createElement(Component, {
      type: resultType,
      size: getIconSize(resultType, size)
    }) : /*#__PURE__*/React.createElement(_spin["default"], null);
    return /*#__PURE__*/React.createElement(Result, (0, _extends2["default"])({}, props, {
      icon: icon,
      status: undefined,
      type: resultType
    }));
  };
};

Result.NetworkError = generatePresetComponent(_NetworkError["default"]);
Result.NoResult = generatePresetComponent(_NoResult["default"]);
Result.NoData = generatePresetComponent(_NoData["default"]);
Result.LoadingFail = generatePresetComponent(_LoadingFail["default"], 'component');
var _default = Result;
exports["default"] = _default;
