// sensd/es/select-panel/Empty.js
import React from 'react';
import LocaleReceiver from '../locale-provider/LocaleReceiver';
import defaultLocale from '../locale/default';
import Result from '../result';
var NoData = Result.NoData,
    NoResult = Result.NoResult;
export var EmptyType;

(function (EmptyType) {
  EmptyType["noData"] = "noData";
  EmptyType["noResult"] = "NoResult";
})(EmptyType || (EmptyType = {}));

var Empty = function Empty(props) {
  var type = props.type,
      noDataDesc = props.noDataDesc,
      noResultDesc = props.noResultDesc,
      noDataTitle = props.noDataTitle,
      noResultTitle = props.noResultTitle;

  var getI18nMsg = function getI18nMsg(name) {
    return /*#__PURE__*/React.createElement(LocaleReceiver, {
      componentName: "SelectPanel",
      defaultLocale: defaultLocale.SelectPanel
    }, function (locale) {
      return locale[name];
    });
  };

  switch (type) {
    case EmptyType.noData:
      return /*#__PURE__*/React.createElement(NoData, {
        title: noDataTitle || getI18nMsg('noData'),
        type: "component",
        subTitle: noDataDesc
      });

    case EmptyType.noResult:
      return /*#__PURE__*/React.createElement(NoResult, {
        title: noResultTitle || getI18nMsg('noResult'),
        type: "component",
        subTitle: noResultDesc || getI18nMsg('noResultDesc')
      });

    default:
      return null;
  }
};

export default Empty;
