// sensd/es/select-panel/List.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React, { useState, useEffect, useMemo, useImperativeHandle, useRef, useContext } from 'react';
import useMergedState from "rc-util/es/hooks/useMergedState";
import { fillFieldNames } from "@sensd/rc-select/es/utils/valueUtil";
import useOptions from "@sensd/rc-select/es/hooks/useOptions";
import SelectPanel from './SelectPanel';
import InternalSelect from './Select';
import { ConfigContext } from '../config-provider';
import Empty, { EmptyType } from './Empty';

var List = function List(props, ref) {
  var propsValue = props.value,
      customizePrefixCls = props.prefixCls,
      defaultValue = props.defaultValue,
      _props$searchValue = props.searchValue,
      searchValue = _props$searchValue === void 0 ? '' : _props$searchValue,
      _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options,
      _props$footer = props.footer,
      footer = _props$footer === void 0 ? true : _props$footer,
      statistics = props.statistics,
      multiple = props.multiple,
      fieldNames = props.fieldNames,
      onChange = props.onChange,
      noDataTitle = props.noDataTitle,
      noDataDesc = props.noDataDesc,
      noResultTitle = props.noResultTitle,
      noResultDesc = props.noResultDesc,
      maxCount = props.maxCount,
      listHeight = props.listHeight,
      showSearch = props.showSearch,
      selectAllText = props.selectAllText,
      others = __rest(props, ["value", "prefixCls", "defaultValue", "searchValue", "options", "footer", "statistics", "multiple", "fieldNames", "onChange", "noDataTitle", "noDataDesc", "noResultTitle", "noResultDesc", "maxCount", "listHeight", "showSearch", "selectAllText"]); // 搜索


  var _useState = useState(searchValue),
      _useState2 = _slicedToArray(_useState, 2),
      query = _useState2[0],
      setQuery = _useState2[1];

  useEffect(function () {
    setQuery(searchValue);
  }, [searchValue]);

  var _useContext = useContext(ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('select-panel', customizePrefixCls); // =========================== Values ===========================

  var _useMergedState = useMergedState(defaultValue, {
    value: propsValue
  }),
      _useMergedState2 = _slicedToArray(_useMergedState, 2),
      internalValue = _useMergedState2[0],
      setInternalValue = _useMergedState2[1];

  useEffect(function () {
    setInternalValue(propsValue);
  }, [propsValue]); // 半选

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      indeterminate = _useState4[0],
      setIndeterminate = _useState4[1]; // 全选


  var _useState5 = useState(false),
      _useState6 = _slicedToArray(_useState5, 2),
      checkAll = _useState6[0],
      setCheckAll = _useState6[1];

  var _useState7 = useState(false),
      _useState8 = _slicedToArray(_useState7, 2),
      hasSelectAllTip = _useState8[0],
      setSelectAllTip = _useState8[1];

  var mergedFieldNames = useMemo(function () {
    return fillFieldNames(fieldNames, false);
  }, [JSON.stringify(fieldNames)]);

  var _useState9 = useState([]),
      _useState10 = _slicedToArray(_useState9, 2),
      flattenOptions = _useState10[0],
      setFlattenOptions = _useState10[1];

  var onListChange = function onListChange(opts) {
    setFlattenOptions(opts);
  };

  var _useOptions = useOptions(options, null, mergedFieldNames),
      valueOptions = _useOptions.valueOptions,
      mergedOptions = _useOptions.options;

  var onCheckboxChange = function onCheckboxChange(value) {
    var disabledSelectedValues = (value || []).filter(function (val) {
      var opt = valueOptions.get(val);
      return opt === null || opt === void 0 ? void 0 : opt.disabled;
    });
    var opts = mergedOptions;
    var data = opts.filter(function (item) {
      return !item.disabled;
    }).map(function (item) {
      return item.value;
    });

    if (maxCount !== undefined) {
      if (maxCount > disabledSelectedValues.length) {
        var enableValue = value.filter(function (val) {
          return !disabledSelectedValues.includes(val);
        });
        var filterData = Array.from(new Set([].concat(_toConsumableArray(data), _toConsumableArray(enableValue))));
        var dataLength = maxCount - disabledSelectedValues.length - enableValue.length;
        data = dataLength > 0 ? filterData.slice(0, dataLength).concat(enableValue) : [];
      } else {
        data = [];
      }
    }

    return {
      filteredValues: [].concat(_toConsumableArray(disabledSelectedValues), _toConsumableArray(data)),
      disabledSelectedValues: disabledSelectedValues
    };
  };

  var handleCheckAll = function handleCheckAll(e) {
    var checked = e.target.checked;
    setCheckAll(checked);

    var _onCheckboxChange = onCheckboxChange(internalValue),
        filteredValues = _onCheckboxChange.filteredValues,
        disabledSelectedValues = _onCheckboxChange.disabledSelectedValues;

    var values = checked ? filteredValues : disabledSelectedValues;
    setInternalValue(values);
    var opts = values.map(function (val) {
      return valueOptions.get(val);
    });
    onChange === null || onChange === void 0 ? void 0 : onChange(values, opts);
  };

  var count = useMemo(function () {
    return flattenOptions.filter(function (item) {
      return !item.group;
    }).length;
  }, [flattenOptions]);

  var handleChange = function handleChange(value, opts) {
    setInternalValue(value);
    onChange === null || onChange === void 0 ? void 0 : onChange(value, opts);
  };

  useEffect(function () {
    if (multiple) {
      // 半选
      setIndeterminate(!!(internalValue === null || internalValue === void 0 ? void 0 : internalValue.length) && (internalValue === null || internalValue === void 0 ? void 0 : internalValue.length) < mergedOptions.length); // 全选

      var _onCheckboxChange2 = onCheckboxChange(internalValue),
          filteredValues = _onCheckboxChange2.filteredValues;

      setCheckAll((internalValue === null || internalValue === void 0 ? void 0 : internalValue.length) === filteredValues.length);
      setSelectAllTip(mergedOptions.some(function (item) {
        return item.disabled === true;
      }));
    }
  }, [internalValue, mergedOptions]);

  var _useMemo = useMemo(function () {
    return [/*#__PURE__*/React.createElement("div", {
      key: EmptyType.noData,
      className: "".concat(prefixCls, "-nodata")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noData,
      noDataTitle: noDataTitle,
      noDataDesc: noDataDesc
    })), /*#__PURE__*/React.createElement("div", {
      key: EmptyType.noResult,
      className: "".concat(prefixCls, "-no-result")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noResult,
      noResultTitle: noResultTitle,
      noResultDesc: noResultDesc
    }))];
  }, [noDataDesc]),
      _useMemo2 = _slicedToArray(_useMemo, 2),
      noDataComp = _useMemo2[0],
      noResultComp = _useMemo2[1];

  var selectRef = useRef(null);
  useImperativeHandle(ref, function () {
    return selectRef.current;
  });
  return /*#__PURE__*/React.createElement(SelectPanel, {
    query: query,
    count: count,
    onSearch: setQuery,
    indeterminate: indeterminate,
    checkAll: checkAll,
    onCheckAll: handleCheckAll,
    footer: footer,
    statistics: statistics,
    multiple: multiple,
    customContent: options.length === 0 && noDataComp,
    noResult: query && count === 0 && noResultComp,
    contentStyle: {
      minHeight: listHeight
    },
    showSearch: showSearch,
    selectAllText: selectAllText,
    hasSelectAllTip: hasSelectAllTip
  }, /*#__PURE__*/React.createElement(InternalSelect, _extends({
    ref: selectRef,
    options: mergedOptions,
    query: query,
    queryString: query,
    onListChange: onListChange,
    multiple: multiple,
    fieldNames: fieldNames,
    valueOptions: valueOptions,
    value: internalValue,
    onChange: handleChange,
    maxCount: maxCount,
    listHeight: listHeight
  }, others)));
};

export default /*#__PURE__*/React.forwardRef(List);
