// sensd/es/select-panel/Select.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import React, { useEffect, useContext, useMemo, useRef, useImperativeHandle } from 'react';
import useOptions from "@sensd/rc-select/es/hooks/useOptions";
import classNames from 'classnames';
import { flattenOptions, fillFieldNames } from "@sensd/rc-select/es/utils/valueUtil";
import useMergedState from "rc-util/es/hooks/useMergedState";
import omit from "rc-util/es/omit";
import List from 'rc-virtual-list';
import { ConfigContext } from '../config-provider';
import SelectOption from './SelectOption';
import searchByKeywords from './utils/searchWords';

var Select = function Select(props, ref) {
  var _props$listHeight = props.listHeight,
      listHeight = _props$listHeight === void 0 ? 358 : _props$listHeight,
      customizePrefixCls = props.prefixCls,
      _props$query = props.query,
      query = _props$query === void 0 ? '' : _props$query,
      _props$queryString = props.queryString,
      queryString = _props$queryString === void 0 ? '' : _props$queryString,
      multiple = props.multiple,
      disabled = props.disabled,
      onChange = props.onChange,
      _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options,
      _props$selectGroupTyp = props.selectGroupType,
      selectGroupType = _props$selectGroupTyp === void 0 ? 'box' : _props$selectGroupTyp,
      defaultValue = props.defaultValue,
      propsValue = props.value,
      onListChange = props.onListChange,
      valueOptions = props.valueOptions,
      fieldNames = props.fieldNames,
      onClick = props.onClick,
      maxCount = props.maxCount,
      _props$listItemHeight = props.listItemHeight,
      listItemHeight = _props$listItemHeight === void 0 ? 34 : _props$listItemHeight,
      _props$virtual = props.virtual,
      virtual = _props$virtual === void 0 ? true : _props$virtual; // =========================== Values ===========================

  var _useMergedState = useMergedState(defaultValue, {
    value: propsValue
  }),
      _useMergedState2 = _slicedToArray(_useMergedState, 2),
      internalValue = _useMergedState2[0],
      setInternalValue = _useMergedState2[1];

  useEffect(function () {
    setInternalValue(propsValue);
  }, [propsValue]);

  var _useContext = useContext(ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('select-panel', customizePrefixCls);
  var listRef = useRef(null);
  var mergedFieldNames = useMemo(function () {
    return fillFieldNames(fieldNames, false);
  }, [JSON.stringify(fieldNames)]);
  var filteredData = useMemo(function () {
    var data = options;

    if (query) {
      data = options.map(function (item) {
        // 如果有分组，只需要搜索分组
        if (item[mergedFieldNames.options]) {
          var filteredChildren = item[mergedFieldNames.options].filter(function (opt) {
            return typeof opt[mergedFieldNames.label] === 'string' && opt[mergedFieldNames.label].toString().includes(query);
          });
          return _extends(_extends({}, item), {
            options: filteredChildren
          });
        }

        return item;
      }).filter(function (item) {
        var childrenOpts = item[mergedFieldNames.options];
        var label = item[mergedFieldNames.label];

        if (!query) {
          return true;
        } // 如果有分组


        if (childrenOpts) {
          return childrenOpts.length > 0;
        } // 搜索label


        if (label && typeof label === 'string') {
          return label.includes(query);
        }

        return false;
      });
    }

    return data;
  }, [query, options]);
  var memoFlattenOptions = useMemo(function () {
    var opts = flattenOptions(filteredData, {
      fieldNames: fieldNames,
      childrenAsData: false
    });
    onListChange === null || onListChange === void 0 ? void 0 : onListChange(opts);
    return opts;
  }, [filteredData]);
  var itemPrefixCls = "".concat(prefixCls, "-option");

  var _useOptions = useOptions(options, null, mergedFieldNames),
      innerValOptions = _useOptions.valueOptions;

  var handleClick = function handleClick(item) {
    if (multiple) {
      // 多选
      var originVal = internalValue || [];
      var newVal;

      if (originVal.includes(item[mergedFieldNames.value])) {
        newVal = originVal.filter(function (val) {
          return val !== item[mergedFieldNames.value];
        });
        setInternalValue(newVal);
        onClick === null || onClick === void 0 ? void 0 : onClick(item[mergedFieldNames.value], false);
      } else {
        newVal = [].concat(_toConsumableArray(originVal), [item[mergedFieldNames.value]]);
        setInternalValue(newVal);
        onClick === null || onClick === void 0 ? void 0 : onClick(item[mergedFieldNames.value], true);
      }

      var opts = newVal.map(function (val) {
        return (valueOptions === null || valueOptions === void 0 ? void 0 : valueOptions.get(val)) || innerValOptions.get(val);
      });
      onChange === null || onChange === void 0 ? void 0 : onChange(newVal, opts);
    } else {
      // 单选
      setInternalValue(item[mergedFieldNames.value]);
      onChange === null || onChange === void 0 ? void 0 : onChange(item[mergedFieldNames.value], item);
      onClick === null || onClick === void 0 ? void 0 : onClick(item[mergedFieldNames.value], internalValue === item[mergedFieldNames.value]);
    }
  };

  useImperativeHandle(ref, function () {
    var _a;

    return {
      scrollTo: (_a = listRef.current) === null || _a === void 0 ? void 0 : _a.scrollTo
    };
  });
  return /*#__PURE__*/React.createElement(List, {
    ref: listRef,
    itemKey: "key",
    data: memoFlattenOptions,
    height: listHeight,
    itemHeight: listItemHeight,
    fullHeight: false,
    virtual: virtual
  }, function (item, itemIndex) {
    var group = item.group,
        data = item.data,
        label = item.label,
        value = item.value;

    var _a = omit(data, ['prefixCls', 'onClick', 'label']),
        key = _a.key,
        optDisabled = _a.disabled,
        others = __rest(_a, ["key", "disabled"]); // Group


    if (group) {
      return selectGroupType === 'line' ? /*#__PURE__*/React.createElement("div", {
        className: classNames(_defineProperty({}, "".concat(itemPrefixCls, "-group-line"), itemIndex !== 0))
      }) : /*#__PURE__*/React.createElement("div", {
        className: classNames("".concat(itemPrefixCls, "-group"))
      }, label !== undefined ? label : key);
    }

    var mergedDisabled = disabled || optDisabled;

    if (maxCount !== undefined && multiple && Array.isArray(internalValue) && internalValue.length >= maxCount && !internalValue.includes(value)) {
      mergedDisabled = true;
    }

    var content = typeof label === 'string' ? searchByKeywords(label, queryString, prefixCls) : label; // Option

    return /*#__PURE__*/React.createElement(SelectOption, _extends({
      multiple: multiple,
      onClick: function onClick() {
        return !mergedDisabled && handleClick(data);
      },
      key: itemIndex,
      prefixCls: itemPrefixCls,
      active: multiple ? internalValue === null || internalValue === void 0 ? void 0 : internalValue.includes(value) : internalValue === value,
      disabled: mergedDisabled,
      label: content
    }, others));
  });
};

export default /*#__PURE__*/React.forwardRef(Select);
