// sensd/es/pagination/Pagination.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _typeof from "@babel/runtime/helpers/esm/typeof";
import _extends from "@babel/runtime/helpers/esm/extends";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import * as React from 'react';
import RcPagination from 'rc-pagination';
import classNames from 'classnames';
import LeftOutlined from "@sensd/icons/es/icons/LeftOutlined";
import RightOutlined from "@sensd/icons/es/icons/RightOutlined";
import MoreOutlined from "@sensd/icons/es/icons/MoreOutlined";
import ArrowDoubuleLeftOutlined from "@sensd/icons/es/icons/ArrowDoubuleLeftOutlined";
import ArrowDoubuleRightOutlined from "@sensd/icons/es/icons/ArrowDoubuleRightOutlined";
import MiniSelect from './MiniSelect';
import Select from '../select';
import LocaleReceiver from '../locale-provider/LocaleReceiver';
import { ConfigContext } from '../config-provider';
import useBreakpoint from '../grid/hooks/useBreakpoint';
import Tooltip from '../tooltip';

var Pagination = function Pagination(_a) {
  var customizePrefixCls = _a.prefixCls,
      customizeSelectPrefixCls = _a.selectPrefixCls,
      customShowTotal = _a.showTotal,
      className = _a.className,
      size = _a.size,
      customLocale = _a.locale,
      selectComponentClass = _a.selectComponentClass,
      _a$showQuickJumper = _a.showQuickJumper,
      showQuickJumper = _a$showQuickJumper === void 0 ? true : _a$showQuickJumper,
      _a$showSizeChanger = _a.showSizeChanger,
      showSizeChanger = _a$showSizeChanger === void 0 ? true : _a$showSizeChanger,
      _a$tooltip = _a.tooltip,
      customTooltipProps = _a$tooltip === void 0 ? true : _a$tooltip,
      restProps = __rest(_a, ["prefixCls", "selectPrefixCls", "showTotal", "className", "size", "locale", "selectComponentClass", "showQuickJumper", "showSizeChanger", "tooltip"]);

  var _useBreakpoint = useBreakpoint(),
      xs = _useBreakpoint.xs;

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls,
      direction = _React$useContext.direction;

  var prefixCls = getPrefixCls('pagination', customizePrefixCls);
  var prevIconRef = /*#__PURE__*/React.createRef();
  var nextIconRef = /*#__PURE__*/React.createRef();

  var getDisabled = function getDisabled(node) {
    var _a, _b, _c;

    return ((_c = (_b = (_a = node === null || node === void 0 ? void 0 : node.current) === null || _a === void 0 ? void 0 : _a.parentNode) === null || _b === void 0 ? void 0 : _b.getAttribute) === null || _c === void 0 ? void 0 : _c.call(_b, 'aria-disabled')) === 'true';
  }; // Tooltip 公共配置


  var tooltipProps = _extends({
    // customTooltipProps 为 false 或整体 disabled 时不显示
    visible: customTooltipProps === false || (restProps === null || restProps === void 0 ? void 0 : restProps.disabled) ? false : undefined,
    arrowPointAtCenter: true,
    mouseLeaveDelay: 0,
    zIndex: 1029
  }, _typeof(customTooltipProps) === 'object' ? customTooltipProps : {});

  var getIconsProps = function getIconsProps(locale) {
    var _a, _b, _c, _d;

    var ellipsis = /*#__PURE__*/React.createElement("span", {
      className: "".concat(prefixCls, "-item-ellipsis")
    }, /*#__PURE__*/React.createElement(MoreOutlined, null));
    var prevTip = [locale === null || locale === void 0 ? void 0 : locale.prev_page, locale === null || locale === void 0 ? void 0 : locale.prev_disable];
    var nextTip = [locale === null || locale === void 0 ? void 0 : locale.next_page, locale === null || locale === void 0 ? void 0 : locale.next_disable];
    var jumpPrevTip = (restProps === null || restProps === void 0 ? void 0 : restProps.showLessItems) ? locale === null || locale === void 0 ? void 0 : locale.prev_3 : locale === null || locale === void 0 ? void 0 : locale.prev_5;
    var jumpNextTip = (restProps === null || restProps === void 0 ? void 0 : restProps.showLessItems) ? locale === null || locale === void 0 ? void 0 : locale.next_3 : locale === null || locale === void 0 ? void 0 : locale.next_5;

    if (direction === 'rtl') {
      var _ref = [nextTip, prevTip];
      prevTip = _ref[0];
      nextTip = _ref[1];
      var _ref2 = [jumpNextTip, jumpPrevTip];
      jumpPrevTip = _ref2[0];
      jumpNextTip = _ref2[1];
    }

    var prevIcon = /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
      title: function title() {
        return prevTip[+getDisabled(prevIconRef)];
      }
    }, (_a = customTooltipProps) === null || _a === void 0 ? void 0 : _a.p_prev), /*#__PURE__*/React.createElement("button", {
      ref: prevIconRef,
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1,
      title: ""
    }, /*#__PURE__*/React.createElement(LeftOutlined, null)));
    var nextIcon = /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
      title: function title() {
        return nextTip[+getDisabled(nextIconRef)];
      }
    }, (_b = customTooltipProps) === null || _b === void 0 ? void 0 : _b.p_next), /*#__PURE__*/React.createElement("button", {
      ref: nextIconRef,
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1,
      title: ""
    }, /*#__PURE__*/React.createElement(RightOutlined, null)));
    var jumpPrevIcon = /*#__PURE__*/React.createElement("a", {
      className: "".concat(prefixCls, "-item-link"),
      title: ""
    }, /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
      title: jumpPrevTip
    }, (_c = customTooltipProps) === null || _c === void 0 ? void 0 : _c.p_jumpPrev), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /*#__PURE__*/React.createElement(ArrowDoubuleLeftOutlined, {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis)));
    var jumpNextIcon = /*#__PURE__*/React.createElement("a", {
      className: "".concat(prefixCls, "-item-link"),
      title: ""
    }, /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
      title: jumpNextTip
    }, (_d = customTooltipProps) === null || _d === void 0 ? void 0 : _d.p_jumpNext), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /*#__PURE__*/React.createElement(ArrowDoubuleRightOutlined, {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis))); // change arrows direction in right-to-left direction

    if (direction === 'rtl') {
      var _ref3 = [nextIcon, prevIcon];
      prevIcon = _ref3[0];
      nextIcon = _ref3[1];
      var _ref4 = [jumpNextIcon, jumpPrevIcon];
      jumpPrevIcon = _ref4[0];
      jumpNextIcon = _ref4[1];
    }

    return {
      prevIcon: prevIcon,
      nextIcon: nextIcon,
      jumpPrevIcon: jumpPrevIcon,
      jumpNextIcon: jumpNextIcon
    };
  };

  var renderPagination = function renderPagination(contextLocale) {
    var locale = _extends(_extends({}, contextLocale), customLocale);

    var isSmall = size === 'small' || !!(xs && !size && restProps.responsive);
    var selectPrefixCls = getPrefixCls('select', customizeSelectPrefixCls);
    var extendedClassName = classNames(_defineProperty({
      mini: isSmall
    }, "".concat(prefixCls, "-rtl"), direction === 'rtl'), className);

    var defaultShowTotal = function defaultShowTotal(_total, range) {
      var _a;

      return (_a = locale === null || locale === void 0 ? void 0 : locale.page_show) === null || _a === void 0 ? void 0 : _a.replace(/\${num}/g, "".concat(range[0], "-").concat(range[1]));
    };

    var defaultSelectComponent = function defaultSelectComponent(props) {
      var _a;

      var allProps = _extends(_extends({}, props), {
        getPopupContainer: function getPopupContainer(node) {
          return node.parentElement || document.body;
        },
        listHeight: 176
      });

      return /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
        title: locale === null || locale === void 0 ? void 0 : locale.page_select,
        getPopupContainer: function getPopupContainer(node) {
          return node;
        }
      }, (_a = customTooltipProps) === null || _a === void 0 ? void 0 : _a.p_select), isSmall ? /*#__PURE__*/React.createElement(MiniSelect, allProps) : /*#__PURE__*/React.createElement(Select, allProps));
    };

    defaultSelectComponent.Option = Select.Option;
    return /*#__PURE__*/React.createElement(RcPagination, _extends({}, getIconsProps(locale), restProps, {
      showQuickJumper: showQuickJumper,
      showSizeChanger: showSizeChanger,
      showTotal: customShowTotal !== null && customShowTotal !== void 0 ? customShowTotal : defaultShowTotal,
      prefixCls: prefixCls,
      selectPrefixCls: selectPrefixCls,
      className: extendedClassName,
      selectComponentClass: selectComponentClass || defaultSelectComponent,
      locale: locale
    }));
  };

  return /*#__PURE__*/React.createElement(LocaleReceiver, {
    componentName: "Pagination"
  }, renderPagination);
};

export default Pagination;
