// sensd/lib/pagination/Pagination.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _rcPagination = _interopRequireDefault(require("rc-pagination"));

var _classnames = _interopRequireDefault(require("classnames"));

var _LeftOutlined = _interopRequireDefault(require("@sensd/icons/LeftOutlined"));

var _RightOutlined = _interopRequireDefault(require("@sensd/icons/RightOutlined"));

var _MoreOutlined = _interopRequireDefault(require("@sensd/icons/MoreOutlined"));

var _ArrowDoubuleLeftOutlined = _interopRequireDefault(require("@sensd/icons/ArrowDoubuleLeftOutlined"));

var _ArrowDoubuleRightOutlined = _interopRequireDefault(require("@sensd/icons/ArrowDoubuleRightOutlined"));

var _MiniSelect = _interopRequireDefault(require("./MiniSelect"));

var _select = _interopRequireDefault(require("../select"));

var _LocaleReceiver = _interopRequireDefault(require("../locale-provider/LocaleReceiver"));

var _configProvider = require("../config-provider");

var _useBreakpoint2 = _interopRequireDefault(require("../grid/hooks/useBreakpoint"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var Pagination = function Pagination(_a) {
  var customizePrefixCls = _a.prefixCls,
      customizeSelectPrefixCls = _a.selectPrefixCls,
      customShowTotal = _a.showTotal,
      className = _a.className,
      size = _a.size,
      customLocale = _a.locale,
      selectComponentClass = _a.selectComponentClass,
      _a$showQuickJumper = _a.showQuickJumper,
      showQuickJumper = _a$showQuickJumper === void 0 ? true : _a$showQuickJumper,
      _a$showSizeChanger = _a.showSizeChanger,
      showSizeChanger = _a$showSizeChanger === void 0 ? true : _a$showSizeChanger,
      _a$tooltip = _a.tooltip,
      customTooltipProps = _a$tooltip === void 0 ? true : _a$tooltip,
      restProps = __rest(_a, ["prefixCls", "selectPrefixCls", "showTotal", "className", "size", "locale", "selectComponentClass", "showQuickJumper", "showSizeChanger", "tooltip"]);

  var _useBreakpoint = (0, _useBreakpoint2["default"])(),
      xs = _useBreakpoint.xs;

  var _React$useContext = React.useContext(_configProvider.ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls,
      direction = _React$useContext.direction;

  var prefixCls = getPrefixCls('pagination', customizePrefixCls);
  var prevIconRef = /*#__PURE__*/React.createRef();
  var nextIconRef = /*#__PURE__*/React.createRef();

  var getDisabled = function getDisabled(node) {
    var _a, _b, _c;

    return ((_c = (_b = (_a = node === null || node === void 0 ? void 0 : node.current) === null || _a === void 0 ? void 0 : _a.parentNode) === null || _b === void 0 ? void 0 : _b.getAttribute) === null || _c === void 0 ? void 0 : _c.call(_b, 'aria-disabled')) === 'true';
  }; // Tooltip 公共配置


  var tooltipProps = (0, _extends2["default"])({
    // customTooltipProps 为 false 或整体 disabled 时不显示
    visible: customTooltipProps === false || (restProps === null || restProps === void 0 ? void 0 : restProps.disabled) ? false : undefined,
    arrowPointAtCenter: true,
    mouseLeaveDelay: 0,
    zIndex: 1029
  }, (0, _typeof2["default"])(customTooltipProps) === 'object' ? customTooltipProps : {});

  var getIconsProps = function getIconsProps(locale) {
    var _a, _b, _c, _d;

    var ellipsis = /*#__PURE__*/React.createElement("span", {
      className: "".concat(prefixCls, "-item-ellipsis")
    }, /*#__PURE__*/React.createElement(_MoreOutlined["default"], null));
    var prevTip = [locale === null || locale === void 0 ? void 0 : locale.prev_page, locale === null || locale === void 0 ? void 0 : locale.prev_disable];
    var nextTip = [locale === null || locale === void 0 ? void 0 : locale.next_page, locale === null || locale === void 0 ? void 0 : locale.next_disable];
    var jumpPrevTip = (restProps === null || restProps === void 0 ? void 0 : restProps.showLessItems) ? locale === null || locale === void 0 ? void 0 : locale.prev_3 : locale === null || locale === void 0 ? void 0 : locale.prev_5;
    var jumpNextTip = (restProps === null || restProps === void 0 ? void 0 : restProps.showLessItems) ? locale === null || locale === void 0 ? void 0 : locale.next_3 : locale === null || locale === void 0 ? void 0 : locale.next_5;

    if (direction === 'rtl') {
      var _ref = [nextTip, prevTip];
      prevTip = _ref[0];
      nextTip = _ref[1];
      var _ref2 = [jumpNextTip, jumpPrevTip];
      jumpPrevTip = _ref2[0];
      jumpNextTip = _ref2[1];
    }

    var prevIcon = /*#__PURE__*/React.createElement(_tooltip["default"], (0, _extends2["default"])({}, tooltipProps, {
      title: function title() {
        return prevTip[+getDisabled(prevIconRef)];
      }
    }, (_a = customTooltipProps) === null || _a === void 0 ? void 0 : _a.p_prev), /*#__PURE__*/React.createElement("button", {
      ref: prevIconRef,
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1,
      title: ""
    }, /*#__PURE__*/React.createElement(_LeftOutlined["default"], null)));
    var nextIcon = /*#__PURE__*/React.createElement(_tooltip["default"], (0, _extends2["default"])({}, tooltipProps, {
      title: function title() {
        return nextTip[+getDisabled(nextIconRef)];
      }
    }, (_b = customTooltipProps) === null || _b === void 0 ? void 0 : _b.p_next), /*#__PURE__*/React.createElement("button", {
      ref: nextIconRef,
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1,
      title: ""
    }, /*#__PURE__*/React.createElement(_RightOutlined["default"], null)));
    var jumpPrevIcon = /*#__PURE__*/React.createElement("a", {
      className: "".concat(prefixCls, "-item-link"),
      title: ""
    }, /*#__PURE__*/React.createElement(_tooltip["default"], (0, _extends2["default"])({}, tooltipProps, {
      title: jumpPrevTip
    }, (_c = customTooltipProps) === null || _c === void 0 ? void 0 : _c.p_jumpPrev), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /*#__PURE__*/React.createElement(_ArrowDoubuleLeftOutlined["default"], {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis)));
    var jumpNextIcon = /*#__PURE__*/React.createElement("a", {
      className: "".concat(prefixCls, "-item-link"),
      title: ""
    }, /*#__PURE__*/React.createElement(_tooltip["default"], (0, _extends2["default"])({}, tooltipProps, {
      title: jumpNextTip
    }, (_d = customTooltipProps) === null || _d === void 0 ? void 0 : _d.p_jumpNext), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /*#__PURE__*/React.createElement(_ArrowDoubuleRightOutlined["default"], {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis))); // change arrows direction in right-to-left direction

    if (direction === 'rtl') {
      var _ref3 = [nextIcon, prevIcon];
      prevIcon = _ref3[0];
      nextIcon = _ref3[1];
      var _ref4 = [jumpNextIcon, jumpPrevIcon];
      jumpPrevIcon = _ref4[0];
      jumpNextIcon = _ref4[1];
    }

    return {
      prevIcon: prevIcon,
      nextIcon: nextIcon,
      jumpPrevIcon: jumpPrevIcon,
      jumpNextIcon: jumpNextIcon
    };
  };

  var renderPagination = function renderPagination(contextLocale) {
    var locale = (0, _extends2["default"])((0, _extends2["default"])({}, contextLocale), customLocale);
    var isSmall = size === 'small' || !!(xs && !size && restProps.responsive);
    var selectPrefixCls = getPrefixCls('select', customizeSelectPrefixCls);
    var extendedClassName = (0, _classnames["default"])((0, _defineProperty2["default"])({
      mini: isSmall
    }, "".concat(prefixCls, "-rtl"), direction === 'rtl'), className);

    var defaultShowTotal = function defaultShowTotal(_total, range) {
      var _a;

      return (_a = locale === null || locale === void 0 ? void 0 : locale.page_show) === null || _a === void 0 ? void 0 : _a.replace(/\${num}/g, "".concat(range[0], "-").concat(range[1]));
    };

    var defaultSelectComponent = function defaultSelectComponent(props) {
      var _a;

      var allProps = (0, _extends2["default"])((0, _extends2["default"])({}, props), {
        getPopupContainer: function getPopupContainer(node) {
          return node.parentElement || document.body;
        },
        listHeight: 176
      });
      return /*#__PURE__*/React.createElement(_tooltip["default"], (0, _extends2["default"])({}, tooltipProps, {
        title: locale === null || locale === void 0 ? void 0 : locale.page_select,
        getPopupContainer: function getPopupContainer(node) {
          return node;
        }
      }, (_a = customTooltipProps) === null || _a === void 0 ? void 0 : _a.p_select), isSmall ? /*#__PURE__*/React.createElement(_MiniSelect["default"], allProps) : /*#__PURE__*/React.createElement(_select["default"], allProps));
    };

    defaultSelectComponent.Option = _select["default"].Option;
    return /*#__PURE__*/React.createElement(_rcPagination["default"], (0, _extends2["default"])({}, getIconsProps(locale), restProps, {
      showQuickJumper: showQuickJumper,
      showSizeChanger: showSizeChanger,
      showTotal: customShowTotal !== null && customShowTotal !== void 0 ? customShowTotal : defaultShowTotal,
      prefixCls: prefixCls,
      selectPrefixCls: selectPrefixCls,
      className: extendedClassName,
      selectComponentClass: selectComponentClass || defaultSelectComponent,
      locale: locale
    }));
  };

  return /*#__PURE__*/React.createElement(_LocaleReceiver["default"], {
    componentName: "Pagination"
  }, renderPagination);
};

var _default = Pagination;
exports["default"] = _default;
