// sensd/es/tree-select/utils.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _typeof from "@babel/runtime/helpers/esm/typeof";
import * as React from 'react';
import { fillLegacyProps } from "@sensd/rc-tree-select/es/utils/legacyUtil";

function getSelectOptionByNode(_ref) {
  var nodeLabel = _ref.nodeLabel,
      nodeValue = _ref.nodeValue,
      nodeDisabled = _ref.nodeDisabled,
      nodeActive = _ref.nodeActive,
      searchValue = _ref.searchValue,
      highlightCls = _ref.highlightCls,
      wrapperLabelCls = _ref.wrapperLabelCls,
      _ref$parentNodeLabelL = _ref.parentNodeLabelList,
      parentNodeLabelList = _ref$parentNodeLabelL === void 0 ? [] : _ref$parentNodeLabelL;
  // 如果 nodeLabel 为 ReactNode，则无法做高亮分词
  var idx = typeof nodeLabel === 'string' ? nodeLabel.toUpperCase().indexOf(searchValue) : -1;
  var highlightLabel = idx > -1 ? /*#__PURE__*/React.createElement("span", {
    className: wrapperLabelCls
  }, nodeLabel.slice(0, idx), /*#__PURE__*/React.createElement("span", {
    className: highlightCls
  }, nodeLabel.slice(idx, idx + searchValue.length)), nodeLabel.slice(idx + searchValue.length)) : /*#__PURE__*/React.createElement("span", {
    className: wrapperLabelCls
  }, nodeLabel);
  var hightlightDetail = /*#__PURE__*/React.createElement(React.Fragment, null, parentNodeLabelList, parentNodeLabelList.length > 0 ? ' / ' : null, highlightLabel);
  return {
    label: highlightLabel,
    value: nodeValue,
    detail: hightlightDetail,
    disabled: nodeDisabled,
    active: nodeActive
  };
}

export function toArray(value) {
  if (Array.isArray(value)) {
    return value;
  }

  return value !== undefined ? [value] : [];
}
export function useFilterTree2List(_ref2) {
  var treeData = _ref2.treeData,
      fieldNames = _ref2.fieldNames,
      highlightCls = _ref2.highlightCls,
      wrapperLabelCls = _ref2.wrapperLabelCls,
      getLabel = _ref2.getLabel,
      multiple = _ref2.multiple,
      treeCheckStrictly = _ref2.treeCheckStrictly,
      treeNodeFilterProp = _ref2.treeNodeFilterProp,
      filterTreeNode = _ref2.filterTreeNode;
  return React.useCallback(function () {
    var internalValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    var searchValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

    if (!searchValue || filterTreeNode === false) {
      return [];
    }

    var upperStr = searchValue.toUpperCase();
    var filterOptionFunc;

    if (typeof filterTreeNode === 'function') {
      filterOptionFunc = filterTreeNode;
    } else {
      filterOptionFunc = function filterOptionFunc(_, dataNode) {
        var value = dataNode[treeNodeFilterProp];
        return String(value).toUpperCase().includes(upperStr);
      };
    }

    var supportEveryNode = multiple && treeCheckStrictly;
    var fieldChildren = fieldNames.children,
        fieldValue = fieldNames.value;
    var result = [];
    var keyValues = toArray(internalValue).map(function (item) {
      return _typeof(item) === 'object' ? item.value : item;
    });

    function dig(list) {
      var parentNodeLabelList = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
      list.forEach(function (dataNode) {
        var legalDataNode = fillLegacyProps(dataNode);
        var label = getLabel(legalDataNode);
        var match = filterOptionFunc === null || filterOptionFunc === void 0 ? void 0 : filterOptionFunc(searchValue, legalDataNode);
        var nodeChild = legalDataNode[fieldChildren];

        if (match && (supportEveryNode || !nodeChild)) {
          var nodeValue = dataNode[fieldValue];
          result.push(getSelectOptionByNode({
            nodeLabel: label,
            nodeValue: nodeValue,
            nodeDisabled: dataNode.disabled,
            nodeActive: keyValues.indexOf(nodeValue) > -1,
            searchValue: upperStr,
            highlightCls: highlightCls,
            wrapperLabelCls: wrapperLabelCls,
            parentNodeLabelList: parentNodeLabelList
          }));
        }

        if (nodeChild) {
          var wrapperLabel = /*#__PURE__*/React.createElement("span", {
            className: wrapperLabelCls
          }, label);
          dig(nodeChild, parentNodeLabelList.length > 0 ? [].concat(_toConsumableArray(parentNodeLabelList), [' / ', wrapperLabel]) : [wrapperLabel]);
        }
      });
    }

    dig(treeData);
    return result;
  }, [treeData, fieldNames, highlightCls, wrapperLabelCls, multiple, treeCheckStrictly, treeNodeFilterProp, filterTreeNode]);
}
export function toLabeledValues(value) {
  var formatValues = toArray(value);
  return formatValues.map(function (val) {
    return {
      value: val
    };
  });
}
