// sensd/es/tree-select/WrapperTree.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import * as React from 'react';
import SelectPanel from '../select-panel';
import Empty, { EmptyType } from '../select-panel/Empty';
import Select from '../select-panel/Select';
import Button from '../button';
import { useFilterTree2List, toArray, toLabeledValues } from './utils';
import LocaleReceiver from '../locale-provider/LocaleReceiver';
import defaultLocale from '../locale/default';

var WrapperTree = function WrapperTree(props) {
  var tree = props.tree,
      prefixCls = props.prefixCls,
      treePrefixCls = props.treePrefixCls,
      _props$treeData = props.treeData,
      treeData = _props$treeData === void 0 ? [] : _props$treeData,
      _props$keyEntities = props.keyEntities,
      keyEntities = _props$keyEntities === void 0 ? {} : _props$keyEntities,
      statistics = props.statistics,
      showCheckAll = props.showCheckAll,
      multiple = props.multiple,
      noDataTitle = props.noDataTitle,
      noDataDesc = props.noDataDesc,
      noResultTitle = props.noResultTitle,
      noResultDesc = props.noResultDesc,
      value = props.value,
      setInnerValue = props.setInnerValue,
      showSearch = props.showSearch,
      defaultSearchValue = props.searchValue,
      onSearch = props.onSearch,
      filterTreeNode = props.filterTreeNode,
      treeCheckStrictly = props.treeCheckStrictly,
      fieldNames = props.fieldNames,
      treeNodeFilterProp = props.treeNodeFilterProp,
      showConfirm = props.showConfirm,
      labelInValue = props.labelInValue,
      dropdownOptLeft = props.dropdownOptLeft,
      dropdownSearchPlaceholder = props.dropdownSearchPlaceholder,
      toggleOpen = props.toggleOpen,
      onSelect = props.onSelect,
      triggerChange = props.triggerChange,
      resetValue = props.resetValue,
      selectAllText = props.selectAllText,
      cancelText = props.cancelText,
      okText = props.okText,
      treeNodeLabelProp = props.treeNodeLabelProp,
      autoClearSearchValue = props.autoClearSearchValue,
      filterSort = props.filterSort;

  var _React$useState = React.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      indeterminate = _React$useState2[0],
      setIndeterminate = _React$useState2[1];

  var _React$useState3 = React.useState(false),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      checkAll = _React$useState4[0],
      setCheckAll = _React$useState4[1];

  var _React$useState5 = React.useState([]),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      searchResult = _React$useState6[0],
      setSearchResult = _React$useState6[1];

  var _React$useState7 = React.useState(defaultSearchValue),
      _React$useState8 = _slicedToArray(_React$useState7, 2),
      searchValue = _React$useState8[0],
      setSearchValue = _React$useState8[1];

  var nodeValueList = React.useMemo(function () {
    if (multiple && treeCheckStrictly) {
      return Object.keys(keyEntities);
    }

    return Object.keys(keyEntities).filter(function (key) {
      return !keyEntities[key].children;
    }).map(function (key) {
      var node = keyEntities[key].node;
      return node[fieldNames.value];
    });
  }, [keyEntities, multiple, treeCheckStrictly]);
  var checkableList = React.useMemo(function () {
    return nodeValueList.filter(function (key) {
      return !keyEntities[key].node.disabled;
    });
  }, [nodeValueList]);
  var getLabel = React.useCallback(function (item) {
    if (item) {
      if (treeNodeLabelProp) {
        return item[treeNodeLabelProp];
      } // Loop from fieldNames


      var titleList = fieldNames._title;

      for (var i = 0; i < titleList.length; i += 1) {
        var title = item[titleList[i]];

        if (title !== undefined) {
          return title;
        }
      }
    }
  }, [fieldNames]);
  var filterTreeData = useFilterTree2List({
    treeData: treeData,
    fieldNames: fieldNames,
    getLabel: getLabel,
    highlightCls: "".concat(prefixCls, "-panel-word-important"),
    wrapperLabelCls: "".concat(treePrefixCls, "-search-split-item"),
    multiple: multiple,
    treeCheckStrictly: treeCheckStrictly,
    treeNodeFilterProp: treeNodeFilterProp,
    filterTreeNode: filterTreeNode
  });
  var getLabelList = React.useCallback(function (rawValue) {
    var formatValues = toArray(rawValue);
    return formatValues.map(function (val) {
      return getLabel(keyEntities[val].node);
    });
  }, [keyEntities]);

  var internalOnCheckAll = function internalOnCheckAll(e) {
    var checked = e.target.checked;
    var disabledSelectedValues = (value || []).filter(function (val) {
      var _a;

      return (_a = keyEntities[val]) === null || _a === void 0 ? void 0 : _a.node.disabled;
    });
    var newValue = checked ? [].concat(_toConsumableArray(disabledSelectedValues), _toConsumableArray(checkableList)) : disabledSelectedValues;
    triggerChange(newValue, {
      selected: checked
    }, 'selection');
  };

  var onInternalSearch = function onInternalSearch(searchText) {
    setSearchValue(searchText);
    var result = filterTreeData(value, searchText);

    if (filterSort) {
      result = result.sort(filterSort);
    }

    setSearchResult(result);
    onSearch === null || onSearch === void 0 ? void 0 : onSearch(searchText);
  };

  var onInternalClickOption = function onInternalClickOption(val, checked) {
    onSelect(val, {
      selected: checked
    });

    if (!multiple) {
      toggleOpen(false);
    }

    if (autoClearSearchValue) {
      setSearchValue('');
    }
  };

  var onInternalOk = function onInternalOk() {
    setInnerValue({
      value: value,
      labelList: labelInValue ? null : getLabelList(value),
      extra: {
        preValue: toLabeledValues(value)
      },
      triggerChange: true
    });
    toggleOpen(false);
  };

  var onInternalCancel = function onInternalCancel() {
    resetValue();
    toggleOpen(false);
  };

  var _React$useMemo = React.useMemo(function () {
    return [/*#__PURE__*/React.createElement("div", {
      key: EmptyType.noData,
      className: "".concat(prefixCls, "-nodata")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noData,
      noDataTitle: noDataTitle,
      noDataDesc: noDataDesc
    })), /*#__PURE__*/React.createElement("div", {
      key: EmptyType.noResult,
      className: "".concat(prefixCls, "-no-result")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noResult,
      noResultTitle: noResultTitle,
      noResultDesc: noResultDesc
    }))];
  }, [noDataDesc]),
      _React$useMemo2 = _slicedToArray(_React$useMemo, 2),
      noDataComp = _React$useMemo2[0],
      noResultComp = _React$useMemo2[1];

  var customRightOpt = /*#__PURE__*/React.createElement(LocaleReceiver, {
    componentName: "Select",
    defaultLocale: defaultLocale.Select
  }, function (locale) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, {
      style: {
        marginRight: 12
      },
      size: "small",
      onClick: onInternalCancel
    }, cancelText || locale.cancelText), /*#__PURE__*/React.createElement(Button, {
      type: "primary",
      size: "small",
      onClick: onInternalOk
    }, okText || locale.okText));
  });
  React.useEffect(function () {
    var _a;

    if (multiple) {
      var checkedLength = (_a = value) === null || _a === void 0 ? void 0 : _a.length;
      setIndeterminate(!!checkedLength && checkedLength < nodeValueList.length);
      setCheckAll(checkedLength >= checkableList.length);
    }

    setSearchResult(filterTreeData(value, searchValue));
  }, [value, checkableList.length, nodeValueList.length, treeCheckStrictly, multiple]);
  return /*#__PURE__*/React.createElement(SelectPanel, {
    showSearch: showSearch,
    query: searchValue,
    onSearch: onInternalSearch,
    indeterminate: indeterminate,
    showCheckAll: typeof showCheckAll === 'boolean' ? showCheckAll : false,
    checkAll: checkAll,
    onCheckAll: internalOnCheckAll,
    multiple: multiple,
    statistics: statistics,
    count: searchValue ? searchResult.length : nodeValueList.length,
    customContent: treeData.length === 0 ? noDataComp : undefined,
    noResult: searchValue && searchResult.length === 0 && noResultComp,
    customRightOpt: showConfirm ? customRightOpt : null,
    customLeftOpt: dropdownOptLeft,
    footer: showCheckAll || showConfirm || dropdownOptLeft,
    placeholder: dropdownSearchPlaceholder,
    selectAllText: selectAllText
  }, searchValue ? /*#__PURE__*/React.createElement(Select, {
    options: searchResult,
    queryString: searchValue,
    multiple: multiple,
    onClick: onInternalClickOption,
    value: value
  }) : tree);
};

export default WrapperTree;
