// sensd/lib/select/option/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _checkbox = _interopRequireDefault(require("../../checkbox"));

var _configProvider = require("../../config-provider");

var _typography = _interopRequireDefault(require("../../typography"));

var _searchWords = _interopRequireDefault(require("../../select-panel/utils/searchWords"));

var Option = function Option(props) {
  var _classNames, _classNames2, _classNames3;

  var prefixIcon = props.prefixIcon,
      suffixIcon = props.suffixIcon,
      multiple = props.multiple,
      disabled = props.disabled,
      checked = props.checked,
      optionTitle = props.optionTitle,
      content = props.content,
      detail = props.detail,
      _props$searchValue = props.searchValue,
      searchValue = _props$searchValue === void 0 ? '' : _props$searchValue;

  var _React$useContext = _react["default"].useContext(_configProvider.ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var prefixCls = getPrefixCls('select-option');
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])((_classNames = {}, (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-wrapper"), true), (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-single"), !multiple), _classNames))
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])((_classNames2 = {}, (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-content"), true), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-has-suffix"), suffixIcon), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-no-suffix"), !suffixIcon && !multiple), _classNames2))
  }, (prefixIcon || multiple) && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefixCls, "-prefix-icon")
  }, multiple && /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
    disabled: disabled,
    checked: checked
  }) || prefixIcon), /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])((_classNames3 = {}, (0, _defineProperty2["default"])(_classNames3, "".concat(prefixCls, "-text"), true), (0, _defineProperty2["default"])(_classNames3, "".concat(prefixCls, "-has-prefix"), prefixIcon || multiple), _classNames3))
  }, /*#__PURE__*/_react["default"].createElement(_typography["default"].Text, {
    ellipsis: {
      tooltip: optionTitle
    },
    className: "".concat(prefixCls, "-ellipsis")
  }, typeof content === 'string' ? (0, _searchWords["default"])(content, searchValue, prefixCls) : content), detail && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefixCls, "-detail")
  }, /*#__PURE__*/_react["default"].createElement(_typography["default"].Text, {
    ellipsis: {
      tooltip: true
    },
    className: "".concat(prefixCls, "-ellipsis")
  }, detail)))), suffixIcon && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefixCls, "-suffix-icon")
  }, suffixIcon));
};

var _default = Option;
exports["default"] = _default;
