// sensd/es/popover/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import * as React from 'react';
import classNames from 'classnames';
import Tooltip from '../tooltip';
import { ConfigContext } from '../config-provider';
import { getRenderPropValue } from '../_util/getRenderPropValue';
import { getTransitionName } from '../_util/motion';
import SizeContext from '../config-provider/SizeContext';
import PopDialog from './PopDialog';
export var popoverBgColor = '#FFFFFF';

var InternalPopover = function InternalPopover(_a, ref) {
  var customizePrefixCls = _a.prefixCls,
      title = _a.title,
      content = _a.content,
      overlayClassName = _a.overlayClassName,
      size = _a.size,
      otherProps = __rest(_a, ["prefixCls", "title", "content", "overlayClassName", "size"]);

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var _size = React.useContext(SizeContext) || size || 'small';

  var getOverlay = function getOverlay(prefixCls) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, title && /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-title")
    }, getRenderPropValue(title)), /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-inner-content")
    }, getRenderPropValue(content)));
  };

  var prefixCls = getPrefixCls('popover', customizePrefixCls);
  var rootPrefixCls = getPrefixCls();
  var customOverlayClassName = classNames(overlayClassName, _defineProperty({}, "".concat(prefixCls, "-").concat(_size), _size));
  return /*#__PURE__*/React.createElement(Tooltip, _extends({
    color: popoverBgColor
  }, otherProps, {
    overlayClassName: customOverlayClassName,
    prefixCls: prefixCls,
    ref: ref,
    overlay: getOverlay(prefixCls),
    transitionName: getTransitionName(rootPrefixCls, 'zoom-big', otherProps.transitionName)
  }));
};

var Popover = /*#__PURE__*/React.forwardRef(InternalPopover);
Popover.displayName = 'Popover';
Popover.defaultProps = {
  placement: 'top',
  trigger: 'click',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1,
  overlayStyle: {}
};
Popover.PopDialog = PopDialog;
export default Popover;
