// sensd/es/popover/PopDialog.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import * as React from 'react';
import classNames from 'classnames';
import Popconfirm from '../popconfirm';
import { ConfigContext } from '../config-provider';
var PopDialog = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var customizePrefixCls = props.prefixCls,
      customOverlayClassName = props.overlayClassName,
      rest = __rest(props, ["prefixCls", "overlayClassName"]);

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var prefixClsDialog = getPrefixCls('popdialog', customizePrefixCls);
  var overlayClassName = classNames(prefixClsDialog, customOverlayClassName, _defineProperty({}, "".concat(prefixClsDialog, "-overflowed"), true));
  return /*#__PURE__*/React.createElement(Popconfirm, _extends({
    okButtonProps: {
      danger: undefined,
      type: 'primary'
    }
  }, rest, {
    overlayClassName: overlayClassName,
    icon: null,
    ref: ref
  }));
});
export default PopDialog;
