// sensd/lib/breadcrumb/BreadcrumbItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _MoreOutlined = _interopRequireDefault(require("@sensd/icons/MoreOutlined"));

var _classnames = _interopRequireDefault(require("classnames"));

var _omit = _interopRequireDefault(require("rc-util/lib/omit"));

var _typography = _interopRequireDefault(require("../typography"));

var _dropdown = _interopRequireDefault(require("../dropdown/dropdown"));

var _configProvider = require("../config-provider");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var BreadcrumbItem = function BreadcrumbItem(_a) {
  var customizePrefixCls = _a.prefixCls,
      _a$separator = _a.separator,
      separator = _a$separator === void 0 ? '/' : _a$separator,
      children = _a.children,
      overlay = _a.overlay,
      dropdownProps = _a.dropdownProps,
      isFoldItem = _a.isFoldItem,
      restProps = __rest(_a, ["prefixCls", "separator", "children", "overlay", "dropdownProps", "isFoldItem"]);

  var _React$useContext = React.useContext(_configProvider.ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var prefixCls = getPrefixCls('breadcrumb', customizePrefixCls);
  /** If overlay is have Wrap a DropDown */

  var renderBreadcrumbNode = function renderBreadcrumbNode(breadcrumbItem) {
    if (overlay) {
      return /*#__PURE__*/React.createElement(_dropdown["default"], (0, _extends2["default"])({
        overlay: overlay,
        placement: "bottomCenter"
      }, dropdownProps), /*#__PURE__*/React.createElement("span", {
        className: (0, _classnames["default"])("".concat(prefixCls, "-overlay-link"), (0, _defineProperty2["default"])({}, "".concat(prefixCls, "-fold-item"), isFoldItem))
      }, breadcrumbItem, /*#__PURE__*/React.createElement(_MoreOutlined["default"], null)));
    }

    return breadcrumbItem;
  };

  var link;

  var href = restProps.href,
      otherProps = __rest(restProps, ["href"]);

  var linkProps = (0, _omit["default"])((0, _extends2["default"])({
    className: "".concat(prefixCls, "-link")
  }, otherProps), ['onClick']);
  link = typeof children === 'string' ? /*#__PURE__*/React.createElement(_typography["default"].Text, (0, _extends2["default"])({}, linkProps, {
    ellipsis: {
      tooltip: /*#__PURE__*/React.createElement("span", {
        className: "".concat(prefixCls, "-tooltip")
      }, children)
    }
  }), children) : /*#__PURE__*/React.createElement("span", linkProps, children);

  if ('href' in restProps) {
    link = /*#__PURE__*/React.createElement("a", {
      className: "".concat(prefixCls, "-typography-link"),
      href: href
    }, link);
  } // wrap to dropDown


  link = renderBreadcrumbNode(link);

  if (children || isFoldItem) {
    return /*#__PURE__*/React.createElement("span", null, link, separator && /*#__PURE__*/React.createElement("span", {
      className: "".concat(prefixCls, "-separator")
    }, separator));
  }

  return null;
};

BreadcrumbItem.__SENS_BREADCRUMB_ITEM = true;
var _default = BreadcrumbItem;
exports["default"] = _default;
