// sensd/lib/cascader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _rcCascader = _interopRequireDefault(require("rc-cascader"));

var _omit = _interopRequireDefault(require("rc-util/lib/omit"));

var _RightFilled = _interopRequireDefault(require("@sensd/icons/RightFilled"));

var _LoadingOutlined = _interopRequireDefault(require("@sensd/icons/LoadingOutlined"));

var _LeftFilled = _interopRequireDefault(require("@sensd/icons/LeftFilled"));

var _devWarning = _interopRequireDefault(require("../_util/devWarning"));

var _configProvider = require("../config-provider");

var _SizeContext = _interopRequireDefault(require("../config-provider/SizeContext"));

var _LocaleReceiver = _interopRequireDefault(require("../locale-provider/LocaleReceiver"));

var _default2 = _interopRequireDefault(require("../locale/default"));

var _iconUtil = _interopRequireDefault(require("../select/utils/iconUtil"));

var _motion = require("../_util/motion");

var _typography = _interopRequireDefault(require("../typography"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _Empty = _interopRequireWildcard(require("../select-panel/Empty"));

var _utils = require("./utils");

var _context = require("../form/context");

var _statusUtils = require("../_util/statusUtils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

function highlightKeyword(str, lowerKeyword, prefixCls) {
  var cells = str.toLowerCase().split(lowerKeyword).reduce(function (list, cur, index) {
    return index === 0 ? [cur] : [].concat((0, _toConsumableArray2["default"])(list), [lowerKeyword, cur]);
  }, []);
  var fillCells = [];
  var start = 0;
  cells.forEach(function (cell, index) {
    var end = start + cell.length;
    var originWorld = str.slice(start, end);
    start = end;

    if (index % 2 === 1) {
      originWorld = /*#__PURE__*/_react["default"].createElement("span", {
        className: "".concat(prefixCls, "-menu-item-keyword"),
        key: "seperator"
      }, originWorld);
    }

    fillCells.push(originWorld);
  });
  return fillCells;
}

var defaultSearchRender = function defaultSearchRender(inputValue, path, prefixCls, fieldNames) {
  var optionList = []; // We do lower here to save perf

  var lower = inputValue.toLowerCase();
  path.forEach(function (node, index) {
    if (index !== 0) {
      optionList.push(' / ');
    }

    var label = node[fieldNames.label];
    var type = (0, _typeof2["default"])(label);

    if (type === 'string' || type === 'number') {
      label = highlightKeyword(String(label), lower, prefixCls);
    }

    optionList.push(label);
  });
  return optionList;
};

var Cascader = /*#__PURE__*/_react["default"].forwardRef(function (props, ref) {
  var _classNames2;

  var customizePrefixCls = props.prefixCls,
      customizeSize = props.size,
      className = props.className,
      multiple = props.multiple,
      _props$bordered = props.bordered,
      bordered = _props$bordered === void 0 ? true : _props$bordered,
      transitionName = props.transitionName,
      _props$choiceTransiti = props.choiceTransitionName,
      choiceTransitionName = _props$choiceTransiti === void 0 ? '' : _props$choiceTransiti,
      popupClassName = props.popupClassName,
      dropdownClassName = props.dropdownClassName,
      expandIcon = props.expandIcon,
      showSearch = props.showSearch,
      _props$allowClear = props.allowClear,
      allowClear = _props$allowClear === void 0 ? true : _props$allowClear,
      notFoundContent = props.notFoundContent,
      direction = props.direction,
      getPopupContainer = props.getPopupContainer,
      children = props.children,
      _props$displayRender = props.displayRender,
      displayRender = _props$displayRender === void 0 ? null : _props$displayRender,
      _props$maxTagPlacehol = props.maxTagPlaceholder,
      maxTagPlaceholder = _props$maxTagPlacehol === void 0 ? null : _props$maxTagPlacehol,
      options = props.options,
      onChange = props.onChange,
      fieldNames = props.fieldNames,
      customStatus = props.status,
      rest = __rest(props, ["prefixCls", "size", "className", "multiple", "bordered", "transitionName", "choiceTransitionName", "popupClassName", "dropdownClassName", "expandIcon", "showSearch", "allowClear", "notFoundContent", "direction", "getPopupContainer", "children", "displayRender", "maxTagPlaceholder", "options", "onChange", "fieldNames", "status"]);

  var restProps = (0, _omit["default"])(rest, ['suffixIcon']);

  var _React$useContext = _react["default"].useContext(_configProvider.ConfigContext),
      getContextPopupContainer = _React$useContext.getPopupContainer,
      getPrefixCls = _React$useContext.getPrefixCls,
      rootDirection = _React$useContext.direction;

  var mergedDirection = direction || rootDirection;
  var isRtl = mergedDirection === 'rtl';

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      dropDownVisible = _useState2[0],
      setDropDownVisible = _useState2[1];

  var selectOptionRef = (0, _react.useRef)([]);

  var getSelectOptionText = function getSelectOptionText(_ref) {
    var selectOption = _ref.selectOption;

    var _a;

    var labelKey = (fieldNames === null || fieldNames === void 0 ? void 0 : fieldNames.label) || 'label';
    var label = selectOption[labelKey]; // 如果 option 有 detail 则 label 为数组: `[label,detail]`

    var labelNode = Array.isArray(label) ? label[0] : label;

    if (labelNode && (labelNode === null || labelNode === void 0 ? void 0 : labelNode.type) === _typography["default"].Text) {
      return (_a = labelNode.props) === null || _a === void 0 ? void 0 : _a.children;
    }

    return labelNode;
  }; // =================== Form =====================


  var _React$useContext2 = _react["default"].useContext(_context.FormItemInputContext),
      contextStatus = _React$useContext2.status,
      hasFeedback = _React$useContext2.hasFeedback,
      feedbackIcon = _React$useContext2.feedbackIcon;

  var mergedStatus = (0, _statusUtils.getMergedStatus)(contextStatus, customStatus); // =================== Warning =====================

  if (process.env.NODE_ENV !== 'production') {
    (0, _devWarning["default"])(popupClassName === undefined, 'Cascader', '`popupClassName` is deprecated. Please use `dropdownClassName` instead.');
    (0, _devWarning["default"])(!multiple || !props.displayRender, 'Cascader', '`displayRender` not work on `multiple`. Please use `tagRender` instead.');
  } // =================== No Found ====================


  var mergedNotFoundContent = notFoundContent || /*#__PURE__*/_react["default"].createElement(_Empty["default"], {
    type: _Empty.EmptyType.noData
  }); // ==================== Prefix =====================


  var rootPrefixCls = getPrefixCls();
  var prefixCls = getPrefixCls('select', customizePrefixCls);
  var cascaderPrefixCls = getPrefixCls('cascader', customizePrefixCls); // =================== Dropdown ====================

  var mergedDropdownClassName = (0, _classnames["default"])(dropdownClassName || popupClassName, "".concat(cascaderPrefixCls, "-dropdown"), (0, _defineProperty2["default"])({}, "".concat(cascaderPrefixCls, "-dropdown-rtl"), mergedDirection === 'rtl')); // ==================== Search =====================

  var mergedShowSearch = _react["default"].useMemo(function () {
    if (!showSearch) {
      return showSearch;
    }

    var searchConfig = {
      render: defaultSearchRender
    };

    if ((0, _typeof2["default"])(showSearch) === 'object') {
      searchConfig = (0, _extends3["default"])((0, _extends3["default"])({}, searchConfig), showSearch);
    }

    return searchConfig;
  }, [showSearch]); // =============== Display Render =================


  var defaultDisplayRender = function defaultDisplayRender(label, selectOption) {
    // fix: value 未匹配到 options 中值
    if (selectOption.every(function (option) {
      return !option;
    })) {
      return (0, _utils.joinNode)(label);
    }

    if (!multiple) {
      var optionTextList = selectOption.map(function (option) {
        return getSelectOptionText({
          selectOption: option
        });
      });
      var mergedLabel = (0, _utils.joinNode)(optionTextList);
      return (0, _utils.wrapChildrenByTypography)(mergedLabel);
    } // 多选


    return (0, _utils.wrapChildrenByTypography)(getSelectOptionText({
      selectOption: selectOption[selectOption.length - 1]
    }));
  };

  var defaultMaxTagPlaceholder = function defaultMaxTagPlaceholder() {
    var selectOption = selectOptionRef.current;

    if (!(selectOption === null || selectOption === void 0 ? void 0 : selectOption.length)) {
      return;
    }

    var displayNodeList = selectOption.map(function (option) {
      return getSelectOptionText({
        selectOption: option[option.length - 1]
      });
    });
    var title = (0, _utils.joinNode)(displayNodeList, '、');
    return /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
      componentName: "Cascader",
      defaultLocale: _default2["default"].Cascader
    }, function (locale) {
      var _a;

      return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
        title: title
      }, locale.selectedCount.replace(/\${count}/, String((_a = selectOptionRef.current) === null || _a === void 0 ? void 0 : _a.length)));
    });
  }; // ===================== Size ======================


  var size = _react["default"].useContext(_SizeContext["default"]);

  var mergedSize = customizeSize || size; // ===================== Icon ======================

  var mergedExpandIcon = expandIcon;

  if (!expandIcon) {
    mergedExpandIcon = isRtl ? /*#__PURE__*/_react["default"].createElement(_LeftFilled["default"], null) : /*#__PURE__*/_react["default"].createElement(_RightFilled["default"], null);
  }

  var loadingIcon = /*#__PURE__*/_react["default"].createElement("span", {
    className: "".concat(prefixCls, "-menu-item-loading-icon")
  }, /*#__PURE__*/_react["default"].createElement(_LoadingOutlined["default"], {
    spin: true
  })); // =================== Multiple ====================


  var checkable = _react["default"].useMemo(function () {
    return multiple ? /*#__PURE__*/_react["default"].createElement("span", {
      className: "".concat(cascaderPrefixCls, "-checkbox-inner")
    }) : false;
  }, [multiple]); // ===================== Icons =====================


  var _getIcons = (0, _iconUtil["default"])((0, _extends3["default"])((0, _extends3["default"])({}, props), {
    hasFeedback: hasFeedback,
    feedbackIcon: feedbackIcon,
    showArrow: props.showArrow,
    multiple: multiple,
    prefixCls: prefixCls,
    dropDownVisible: dropDownVisible
  })),
      suffixIcon = _getIcons.suffixIcon,
      removeIcon = _getIcons.removeIcon,
      clearIcon = _getIcons.clearIcon; // =================== Callback ====================


  var handleChange = function handleChange(value, selectOption) {
    selectOptionRef.current = selectOption;
    onChange === null || onChange === void 0 ? void 0 : onChange(value, selectOption);
  }; // ==================== Render =====================


  var customDisplayRender = null;

  if (!children) {
    customDisplayRender = displayRender || defaultDisplayRender;
  }

  var mergeOptions = function mergeOptions(targetOptions) {
    if (!Array.isArray(targetOptions)) {
      return targetOptions;
    } // eslint-disable-next-line @typescript-eslint/no-shadow


    return targetOptions.map(function (_a) {
      var _extends2;

      var detail = _a.detail,
          rest = __rest(_a, ["detail"]); // fix: fieldNames value 和 label 相同时 不用 Typography.Text 包裹 label
      // 因为 label 和 value 指向的值必须是有效的 value


      var _ref2 = fieldNames || {},
          _ref2$label = _ref2.label,
          targetLabelKey = _ref2$label === void 0 ? 'label' : _ref2$label,
          _ref2$value = _ref2.value,
          targetValueKey = _ref2$value === void 0 ? 'value' : _ref2$value,
          _ref2$children = _ref2.children,
          targetChildrenKey = _ref2$children === void 0 ? 'children' : _ref2$children;

      var label = rest[targetLabelKey];
      var content = rest[targetChildrenKey];
      var isSameFieldKey = targetLabelKey === targetValueKey;
      var wrappedLabel = isSameFieldKey ? label : (0, _utils.wrapChildrenByTypography)(label);

      if (detail && !isSameFieldKey) {
        wrappedLabel = [wrappedLabel,
        /*#__PURE__*/
        // eslint-disable-next-line react/jsx-key
        _react["default"].createElement("div", {
          className: "".concat(cascaderPrefixCls, "-content-detail")
        }, (0, _utils.wrapChildrenByTypography)(detail, {
          className: "".concat(cascaderPrefixCls, "-content-detail-text")
        }))];
      }

      return (0, _extends3["default"])((0, _extends3["default"])({}, rest), (_extends2 = {}, (0, _defineProperty2["default"])(_extends2, targetChildrenKey, content && mergeOptions(content)), (0, _defineProperty2["default"])(_extends2, targetLabelKey, wrappedLabel), _extends2));
    });
  };

  var mergedOptions = options;

  if (!mergedShowSearch) {
    mergedOptions = mergeOptions(options);
  }

  return /*#__PURE__*/_react["default"].createElement(_rcCascader["default"], (0, _extends3["default"])({
    prefixCls: prefixCls,
    className: (0, _classnames["default"])(!customizePrefixCls && cascaderPrefixCls, (_classNames2 = {}, (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-lg"), mergedSize === 'large'), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-sm"), mergedSize === 'small'), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-rtl"), isRtl), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-borderless"), !bordered), _classNames2), (0, _statusUtils.getStatusClassNames)(prefixCls, mergedStatus, hasFeedback), className)
  }, restProps, {
    direction: mergedDirection,
    notFoundContent: mergedNotFoundContent,
    allowClear: allowClear,
    showSearch: mergedShowSearch,
    expandIcon: mergedExpandIcon,
    inputIcon: suffixIcon,
    removeIcon: removeIcon,
    clearIcon: clearIcon,
    loadingIcon: loadingIcon,
    checkable: checkable,
    dropdownClassName: mergedDropdownClassName,
    dropdownPrefixCls: customizePrefixCls || cascaderPrefixCls,
    choiceTransitionName: (0, _motion.getTransitionName)(rootPrefixCls, '', choiceTransitionName),
    transitionName: (0, _motion.getTransitionName)(rootPrefixCls, 'slide-up', transitionName),
    getPopupContainer: getPopupContainer || getContextPopupContainer,
    ref: ref,
    onDropdownVisibleChange: setDropDownVisible,
    displayRender: customDisplayRender,
    maxTagPlaceholder: maxTagPlaceholder || defaultMaxTagPlaceholder,
    options: mergedOptions,
    onChange: handleChange,
    fieldNames: fieldNames
  }), children);
});

Cascader.displayName = 'Cascader';
var _default = Cascader;
exports["default"] = _default;
