// sensd/es/steps/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import * as React from 'react';
import omit from "rc-util/es/omit";
import RcSteps from 'rc-steps';
import CheckOutlined from "@sensd/icons/es/icons/CheckOutlined";
import CloseOutlined from "@sensd/icons/es/icons/CloseOutlined";
import LoadingOutlined from "@sensd/icons/es/icons/LoadingOutlined";
import classNames from 'classnames';
import { ConfigContext } from '../config-provider';
import Progress from '../progress';
import useBreakpoint from '../grid/hooks/useBreakpoint';
import Typography from '../typography';
import Tooltip from '../tooltip';
var Text = Typography.Text;
var RcStep = RcSteps.Step;
var progressDefaultWidth = 40;
var progressSmallWidth = 32;

var Steps = function Steps(props) {
  var _classNames;

  var percent = props.percent,
      size = props.size,
      className = props.className,
      direction = props.direction,
      responsive = props.responsive,
      children = props.children,
      status = props.status;

  var _useBreakpoint = useBreakpoint(),
      xs = _useBreakpoint.xs;

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls,
      rtlDirection = _React$useContext.direction;

  var getDirection = React.useCallback(function () {
    return responsive && xs ? 'vertical' : direction;
  }, [xs, direction]);
  var prefixCls = getPrefixCls('steps', props.prefixCls);
  var iconPrefix = getPrefixCls('', props.iconPrefix);
  var stepsClassName = classNames((_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-rtl"), rtlDirection === 'rtl'), _defineProperty(_classNames, "".concat(prefixCls, "-with-progress"), percent !== undefined), _classNames), className);
  var icons = {
    finish: /*#__PURE__*/React.createElement(CheckOutlined, {
      className: "".concat(prefixCls, "-finish-icon")
    }),
    error: /*#__PURE__*/React.createElement(CloseOutlined, {
      className: "".concat(prefixCls, "-error-icon")
    })
  };

  var stepIconRender = function stepIconRender(_ref) {
    var node = _ref.node,
        StepStatus = _ref.status,
        index = _ref.index;

    var _a, _b;

    if (StepStatus === 'process' && percent !== undefined) {
      // currently it's hard-coded, since we can't easily read the actually width of icon
      var progressWidth = size === 'small' ? progressSmallWidth : progressDefaultWidth;
      var iconWithProgress = /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefixCls, "-progress-icon")
      }, /*#__PURE__*/React.createElement(Progress, {
        type: "circle",
        percent: percent,
        width: progressWidth,
        strokeWidth: 4,
        format: function format() {
          return null;
        }
      }), node);
      return iconWithProgress;
    }

    if (StepStatus === 'error') {
      // status为error则不走原本的icon显示，直接显步骤数
      var stepList = React.Children.toArray(children); // 若是传入icon则不显示步骤数，直接显示 icon

      if (!((_b = (_a = stepList[index]) === null || _a === void 0 ? void 0 : _a.props) === null || _b === void 0 ? void 0 : _b.icon)) {
        return /*#__PURE__*/React.createElement("span", {
          className: "".concat(prefixCls, "-icon")
        }, index + 1);
      }
    }

    if (StepStatus === 'loading') {
      return /*#__PURE__*/React.createElement(LoadingOutlined, {
        className: "".concat(prefixCls, "-loading-icon")
      });
    }

    return node;
  };

  return /*#__PURE__*/React.createElement(RcSteps, _extends({
    icons: icons
  }, omit(props, ['percent', 'responsive']), {
    status: status // 避免ts检查报错
    ,
    direction: getDirection(),
    stepIcon: stepIconRender,
    prefixCls: prefixCls,
    iconPrefix: iconPrefix,
    className: stepsClassName
  }));
};

var Step = function Step(props) {
  var title = props.title,
      description = props.description,
      status = props.status,
      tip = props.tip,
      restProps = __rest(props, ["title", "description", "status", "tip"]); // 创建ellipsis组件


  var createEllipsisText = function createEllipsisText(text, showTip) {
    var shouldShowTip = showTip && tip;

    if (typeof text !== 'string') {
      return shouldShowTip ? /*#__PURE__*/React.createElement(Tooltip, {
        title: tip
      }, text) : text;
    }

    var ellipsisText = /*#__PURE__*/React.createElement(Text, {
      ellipsis: {
        tooltip: !shouldShowTip && text
      }
    }, text);
    return shouldShowTip ? /*#__PURE__*/React.createElement(Tooltip, {
      title: shouldShowTip
    }, ellipsisText) : ellipsisText;
  };

  var titleEllipsis = createEllipsisText(title, true);
  var descriptionEllipsis = createEllipsisText(description);
  return /*#__PURE__*/React.createElement(RcStep, _extends({}, restProps, {
    status: status // 避免ts检查报错
    ,
    title: titleEllipsis,
    description: descriptionEllipsis
  }));
};

Steps.Step = Step;
Steps.defaultProps = {
  current: 0,
  responsive: true
};
export default Steps;
