// sensd/lib/select-panel/Empty.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.EmptyType = void 0;

var _react = _interopRequireDefault(require("react"));

var _LocaleReceiver = _interopRequireDefault(require("../locale-provider/LocaleReceiver"));

var _default2 = _interopRequireDefault(require("../locale/default"));

var _result = _interopRequireDefault(require("../result"));

var NoData = _result["default"].NoData,
    NoResult = _result["default"].NoResult;
var EmptyType;
exports.EmptyType = EmptyType;

(function (EmptyType) {
  EmptyType["noData"] = "noData";
  EmptyType["noResult"] = "NoResult";
})(EmptyType || (exports.EmptyType = EmptyType = {}));

var Empty = function Empty(props) {
  var type = props.type,
      noDataDesc = props.noDataDesc,
      noResultDesc = props.noResultDesc,
      noDataTitle = props.noDataTitle,
      noResultTitle = props.noResultTitle;

  var getI18nMsg = function getI18nMsg(name) {
    return /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
      componentName: "SelectPanel",
      defaultLocale: _default2["default"].SelectPanel
    }, function (locale) {
      return locale[name];
    });
  };

  switch (type) {
    case EmptyType.noData:
      return /*#__PURE__*/_react["default"].createElement(NoData, {
        title: noDataTitle || getI18nMsg('noData'),
        type: "component",
        subTitle: noDataDesc
      });

    case EmptyType.noResult:
      return /*#__PURE__*/_react["default"].createElement(NoResult, {
        title: noResultTitle || getI18nMsg('noResult'),
        type: "component",
        subTitle: noResultDesc || getI18nMsg('noResultDesc')
      });

    default:
      return null;
  }
};

var _default = Empty;
exports["default"] = _default;
