// sensd/lib/select-panel/SelectPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _LeftOutlined = _interopRequireDefault(require("@sensd/icons/LeftOutlined"));

var _classnames = _interopRequireDefault(require("classnames"));

var _configProvider = require("../config-provider");

var _LocaleReceiver = _interopRequireDefault(require("../locale-provider/LocaleReceiver"));

var _default2 = _interopRequireDefault(require("../locale/default"));

var _input = _interopRequireDefault(require("../input"));

var _checkbox = _interopRequireDefault(require("../checkbox"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _reactNode = require("../_util/reactNode");

var _divider = _interopRequireDefault(require("../divider"));

var _context = require("../form/context");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var SHOW_TOTAL_MIN_COUNT = 9;

var SelectPanel = function SelectPanel(props) {
  var customizePrefixCls = props.prefixCls,
      _props$showSearch = props.showSearch,
      showSearch = _props$showSearch === void 0 ? false : _props$showSearch,
      searchSuffix = props.searchSuffix,
      _props$statistics = props.statistics,
      statistics = _props$statistics === void 0 ? false : _props$statistics,
      customLeftOpt = props.customLeftOpt,
      customRightOpt = props.customRightOpt,
      multiple = props.multiple,
      _props$showCheckAll = props.showCheckAll,
      showCheckAll = _props$showCheckAll === void 0 ? true : _props$showCheckAll,
      _props$footer = props.footer,
      footer = _props$footer === void 0 ? false : _props$footer,
      children = props.children,
      onCheckAll = props.onCheckAll,
      indeterminate = props.indeterminate,
      checkAll = props.checkAll,
      query = props.query,
      onSearch = props.onSearch,
      _props$count = props.count,
      count = _props$count === void 0 ? 0 : _props$count,
      customContent = props.customContent,
      noResult = props.noResult,
      disableCheckAll = props.disableCheckAll,
      placeholder = props.placeholder,
      selectAllText = props.selectAllText,
      contentStyle = props.contentStyle;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      focused = _useState2[0],
      setFocused = _useState2[1];

  var inputRef = (0, _react.useRef)(null);

  var _useContext = (0, _react.useContext)(_configProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('select-panel', customizePrefixCls);

  var handleBack = function handleBack() {
    var _a;

    onSearch === null || onSearch === void 0 ? void 0 : onSearch('');
    setFocused(true);
    (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
  };

  var handleSearch = function handleSearch(e) {
    var searchVal = e.target.value;
    onSearch === null || onSearch === void 0 ? void 0 : onSearch(searchVal);
  };

  var defaultStatistics = function defaultStatistics(locale) {
    var text = ''; // 搜索无数据，或者总数据小于9条。不显示统计区

    if (count === 0 || !query && count < SHOW_TOTAL_MIN_COUNT) {
      return null;
    }

    if (query) {
      text = locale.searchCount.replace(/\${count}/, count);
    } else {
      text = locale.totalCount.replace(/\${count}/, count);
    }

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-statistics")
    }, text);
  };

  var defaultFooter = function defaultFooter(locale) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-footer")
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-operate-left")
    }, customLeftOpt || multiple && showCheckAll && /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
      indeterminate: indeterminate,
      onChange: onCheckAll,
      checked: checkAll,
      disabled: disableCheckAll
    }, selectAllText || locale.selectAll)), /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-operate-right")
    }, customRightOpt));
  };

  var renderFooter = function renderFooter() {
    if (footer === false || count === 0 || query) return null;

    if (footer !== true) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-footer")
      }, footer);
    }

    return /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
      componentName: "SelectPanel",
      defaultLocale: _default2["default"].SelectPanel
    }, defaultFooter);
  };

  var renderStatistics = function renderStatistics() {
    if (statistics === false) return null;

    if (statistics !== true) {
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-statistics")
      }, statistics);
    }

    return /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
      componentName: "SelectPanel",
      defaultLocale: _default2["default"].SelectPanel
    }, defaultStatistics);
  };

  var backNode = function backNode(locale) {
    return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
      title: locale.back
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-back-icon")
    }, /*#__PURE__*/_react["default"].createElement(_LeftOutlined["default"], {
      onClick: handleBack
    }), /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-back-line")
    })));
  };

  var onKeyDownDisable = function onKeyDownDisable(e) {
    e.stopPropagation();
  };

  if ((0, _reactNode.isValidElement)(customContent)) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: prefixCls
    }, customContent);
  }

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])((0, _defineProperty2["default"])({}, "".concat(prefixCls, "-width-search"), showSearch), prefixCls)
  }, showSearch && /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])((0, _defineProperty2["default"])({}, "".concat(prefixCls, "-search-focused"), focused), "".concat(prefixCls, "-header-search")),
    onKeyDown: onKeyDownDisable
  }, query && /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
    componentName: "SelectPanel",
    defaultLocale: _default2["default"].SelectPanel
  }, backNode), /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
    componentName: "SelectPanel",
    defaultLocale: _default2["default"].global
  }, function (locale) {
    return /*#__PURE__*/_react["default"].createElement(_context.NoFormStyle, {
      override: true,
      status: true
    }, /*#__PURE__*/_react["default"].createElement(_input["default"].Search, {
      ref: inputRef,
      onFocus: function onFocus() {
        return setFocused(true);
      },
      onBlur: function onBlur() {
        return setFocused(false);
      },
      value: query,
      allowClear: true,
      concise: true,
      suffix: searchSuffix && /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_divider["default"], {
        className: "".concat(prefixCls, "-search-divider"),
        type: "vertical"
      }), searchSuffix),
      onChange: handleSearch,
      className: "".concat(prefixCls, "-search-input"),
      placeholder: placeholder || locale.placeholder
    }));
  })), renderStatistics(), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefixCls, "-content"),
    style: contentStyle
  }, children, noResult), renderFooter());
};

var _default = SelectPanel;
exports["default"] = _default;
