// sensd/es/popconfirm/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

var _this = this;

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import * as React from 'react';
import classNames from 'classnames';
import useMergedState from "rc-util/es/hooks/useMergedState";
import WarningFilled from "@sensd/icons/es/icons/WarningFilled";
import KeyCode from "rc-util/es/KeyCode";
import Tooltip from '../tooltip';
import Button from '../button';
import { convertLegacyProps } from '../button/button';
import LocaleReceiver from '../locale-provider/LocaleReceiver';
import defaultLocale from '../locale/default';
import { ConfigContext } from '../config-provider';
import { getRenderPropValue } from '../_util/getRenderPropValue';
import { cloneElement } from '../_util/reactNode';
import { getTransitionName } from '../_util/motion';
import ActionButton from '../_util/ActionButton';
import useDestroyed from '../_util/hooks/useDestroyed';
import SizeContext from '../config-provider/SizeContext';
import { popoverBgColor } from '../popover'; // @privacy

export var OverLay = function OverLay(props) {
  var okButtonProps = props.okButtonProps,
      cancelButtonProps = props.cancelButtonProps,
      title = props.title,
      cancelText = props.cancelText,
      okText = props.okText,
      okType = props.okType,
      icon = props.icon,
      content = props.content,
      _props$showCancel = props.showCancel,
      showCancel = _props$showCancel === void 0 ? true : _props$showCancel,
      prefixCls = props.prefixCls,
      onCancel = props.onCancel,
      close = props.close,
      footer = props.footer,
      onConfirm = props.onConfirm,
      size = props.size,
      popconfirmLocale = props.popconfirmLocale;

  var _React$useContext = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext.getPrefixCls;

  var _React$useState = React.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      contentOverflowed = _React$useState2[0],
      setContentOverflowed = _React$useState2[1];

  var contentRef = React.useRef(null);
  React.useLayoutEffect(function () {
    var node = contentRef.current;
    var parentNode = node === null || node === void 0 ? void 0 : node.parentElement;

    if (((node === null || node === void 0 ? void 0 : node.offsetHeight) || 0) > ((parentNode === null || parentNode === void 0 ? void 0 : parentNode.clientHeight) || 0)) {
      setContentOverflowed(true);
    } else {
      setContentOverflowed(false);
    }
  }, [content]);
  var cls = classNames("".concat(prefixCls, "-inner-content"), _defineProperty({}, "".concat(prefixCls, "-inner-content-overflowed"), contentOverflowed));
  var btnSizeProp = size === 'small' ? {
    size: 'small'
  } : {};
  return /*#__PURE__*/React.createElement("div", {
    className: cls
  }, title && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-message")
  }, icon, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-message-title")
  }, getRenderPropValue(title))), !!content && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-message-content")
  }, /*#__PURE__*/React.createElement("div", {
    ref: contentRef
  }, content)), footer !== null && footer !== void 0 ? footer : /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-buttons")
  }, /*#__PURE__*/React.createElement(ActionButton, {
    buttonProps: _extends(_extends(_extends({
      danger: 'confirm'
    }, btnSizeProp), convertLegacyProps(okType)), okButtonProps),
    actionFn: onConfirm,
    close: close,
    prefixCls: getPrefixCls('btn'),
    quitOnNullishReturnValue: true,
    emitEvent: true
  }, okText || popconfirmLocale.okText), showCancel && /*#__PURE__*/React.createElement(Button, _extends({
    onClick: onCancel
  }, btnSizeProp, cancelButtonProps), cancelText || popconfirmLocale.cancelText)));
};
var Popconfirm = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _React$useContext2 = React.useContext(ConfigContext),
      getPrefixCls = _React$useContext2.getPrefixCls;

  var _useMergedState = useMergedState(false, {
    value: props.visible,
    defaultValue: props.defaultVisible
  }),
      _useMergedState2 = _slicedToArray(_useMergedState, 2),
      visible = _useMergedState2[0],
      setVisible = _useMergedState2[1];

  var size = React.useContext(SizeContext) || props.size || 'small';
  var isDestroyed = useDestroyed();

  var settingVisible = function settingVisible(value, e) {
    var _a;

    if (!isDestroyed()) {
      setVisible(value);
    }

    (_a = props.onVisibleChange) === null || _a === void 0 ? void 0 : _a.call(props, value, e);
  };

  var close = function close(e) {
    settingVisible(false, e);
  };

  var onConfirm = function onConfirm(e) {
    var _a;

    return (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(_this, e);
  };

  var onCancel = function onCancel(e) {
    var _a;

    settingVisible(false, e);
    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(_this, e);
  };

  var _onKeyDown = function onKeyDown(e) {
    if (e.keyCode === KeyCode.ESC && visible) {
      settingVisible(false, e);
    }
  };

  var onVisibleChange = function onVisibleChange(value) {
    var disabled = props.disabled;

    if (disabled) {
      return;
    }

    settingVisible(value);
  };

  var customizePrefixCls = props.prefixCls,
      placement = props.placement,
      children = props.children,
      overlayClassName = props.overlayClassName,
      restProps = __rest(props, ["prefixCls", "placement", "children", "overlayClassName"]);

  var prefixCls = getPrefixCls('popover', customizePrefixCls);
  var prefixClsConfirm = getPrefixCls('popconfirm', customizePrefixCls);
  var overlayClassNames = classNames(prefixClsConfirm, overlayClassName, _defineProperty({}, "".concat(prefixClsConfirm, "-").concat(size), size));
  var overlay = /*#__PURE__*/React.createElement(LocaleReceiver, {
    componentName: "Popconfirm",
    defaultLocale: defaultLocale.Popconfirm
  }, function (popconfirmLocale) {
    return /*#__PURE__*/React.createElement(OverLay, _extends(_extends({
      size: size
    }, props), {
      onConfirm: onConfirm,
      onCancel: onCancel,
      close: close,
      popconfirmLocale: popconfirmLocale,
      prefixCls: prefixCls
    }));
  });
  var rootPrefixCls = getPrefixCls();
  return /*#__PURE__*/React.createElement(Tooltip, _extends({
    color: popoverBgColor
  }, restProps, {
    prefixCls: prefixCls,
    placement: placement,
    onVisibleChange: onVisibleChange,
    visible: visible,
    overlay: overlay,
    overlayClassName: overlayClassNames,
    ref: ref,
    transitionName: getTransitionName(rootPrefixCls, 'zoom-big', props.transitionName)
  }), cloneElement(children, {
    onKeyDown: function onKeyDown(e) {
      var _a, _b;

      if ( /*#__PURE__*/React.isValidElement(children)) {
        (_b = children === null || children === void 0 ? void 0 : (_a = children.props).onKeyDown) === null || _b === void 0 ? void 0 : _b.call(_a, e);
      }

      _onKeyDown(e);
    }
  }));
});
Popconfirm.defaultProps = {
  placement: 'top',
  trigger: 'click',
  okType: 'default',
  icon: /*#__PURE__*/React.createElement(WarningFilled, null),
  disabled: false
};
export default Popconfirm;
