// sensd/lib/tree-select/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TreeNode", {
  enumerable: true,
  get: function get() {
    return _rcTreeSelect.TreeNode;
  }
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _rcTreeSelect = _interopRequireWildcard(require("@sensd/rc-tree-select"));

var _classnames = _interopRequireDefault(require("classnames"));

var _omit = _interopRequireDefault(require("rc-util/lib/omit"));

var _useMergedState3 = _interopRequireDefault(require("rc-util/lib/hooks/useMergedState"));

var _configProvider = require("../config-provider");

var _devWarning = _interopRequireDefault(require("../_util/devWarning"));

var _iconUtil = _interopRequireDefault(require("../select/utils/iconUtil"));

var _iconUtil2 = _interopRequireDefault(require("../tree/utils/iconUtil"));

var _SizeContext = _interopRequireDefault(require("../config-provider/SizeContext"));

var _motion = require("../_util/motion");

var _WrapperTree = _interopRequireDefault(require("./WrapperTree"));

var _Selector = _interopRequireDefault(require("./Selector"));

var _utils = require("./utils");

var _context = require("../form/context");

var _statusUtils = require("../_util/statusUtils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var InternalTreeSelect = function InternalTreeSelect(_a, ref) {
  var _classNames2;

  var customizePrefixCls = _a.prefixCls,
      customizeSize = _a.size,
      _a$bordered = _a.bordered,
      bordered = _a$bordered === void 0 ? true : _a$bordered,
      className = _a.className,
      treeCheckable = _a.treeCheckable,
      _a$treeNodeFilterProp = _a.treeNodeFilterProp,
      treeNodeFilterProp = _a$treeNodeFilterProp === void 0 ? 'title' : _a$treeNodeFilterProp,
      multiple = _a.multiple,
      _a$listHeight = _a.listHeight,
      listHeight = _a$listHeight === void 0 ? 358 : _a$listHeight,
      _a$listItemHeight = _a.listItemHeight,
      listItemHeight = _a$listItemHeight === void 0 ? 26 : _a$listItemHeight,
      notFoundContent = _a.notFoundContent,
      _switcherIcon = _a.switcherIcon,
      treeLine = _a.treeLine,
      getPopupContainer = _a.getPopupContainer,
      dropdownClassName = _a.dropdownClassName,
      _a$treeIcon = _a.treeIcon,
      treeIcon = _a$treeIcon === void 0 ? false : _a$treeIcon,
      transitionName = _a.transitionName,
      _a$choiceTransitionNa = _a.choiceTransitionName,
      choiceTransitionName = _a$choiceTransitionNa === void 0 ? '' : _a$choiceTransitionNa,
      defaultValue = _a.defaultValue,
      propsValue = _a.value,
      onChange = _a.onChange,
      showConfirm = _a.showConfirm,
      showCheckAll = _a.showCheckAll,
      statistics = _a.statistics,
      noDataDesc = _a.noDataDesc,
      noDataTitle = _a.noDataTitle,
      noResultDesc = _a.noResultDesc,
      noResultTitle = _a.noResultTitle,
      showSearch = _a.showSearch,
      searchValue = _a.searchValue,
      onSearch = _a.onSearch,
      onDropdownVisibleChange = _a.onDropdownVisibleChange,
      selectorSimpleMode = _a.selectorSimpleMode,
      selectorRender = _a.selectorRender,
      filterTreeNode = _a.filterTreeNode,
      _a$showArrow = _a.showArrow,
      showArrow = _a$showArrow === void 0 ? true : _a$showArrow,
      placeholder = _a.placeholder,
      maxTagCount = _a.maxTagCount,
      maxTagPlaceholder = _a.maxTagPlaceholder,
      treeCheckStrictly = _a.treeCheckStrictly,
      labelInValue = _a.labelInValue,
      dropdownOptLeft = _a.dropdownOptLeft,
      dropdownSearchPlaceholder = _a.dropdownSearchPlaceholder,
      selectAllText = _a.selectAllText,
      cancelText = _a.cancelText,
      okText = _a.okText,
      prefixIcon = _a.prefixIcon,
      treeNodeLabelProp = _a.treeNodeLabelProp,
      autoClearSearchValue = _a.autoClearSearchValue,
      customStatus = _a.status,
      filterSort = _a.filterSort,
      props = __rest(_a, ["prefixCls", "size", "bordered", "className", "treeCheckable", "treeNodeFilterProp", "multiple", "listHeight", "listItemHeight", "notFoundContent", "switcherIcon", "treeLine", "getPopupContainer", "dropdownClassName", "treeIcon", "transitionName", "choiceTransitionName", "defaultValue", "value", "onChange", "showConfirm", "showCheckAll", "statistics", "noDataDesc", "noDataTitle", "noResultDesc", "noResultTitle", "showSearch", "searchValue", "onSearch", "onDropdownVisibleChange", "selectorSimpleMode", "selectorRender", "filterTreeNode", "showArrow", "placeholder", "maxTagCount", "maxTagPlaceholder", "treeCheckStrictly", "labelInValue", "dropdownOptLeft", "dropdownSearchPlaceholder", "selectAllText", "cancelText", "okText", "prefixIcon", "treeNodeLabelProp", "autoClearSearchValue", "status", "filterSort"]);

  var _React$useContext = React.useContext(_configProvider.ConfigContext),
      getContextPopupContainer = _React$useContext.getPopupContainer,
      getPrefixCls = _React$useContext.getPrefixCls,
      renderEmpty = _React$useContext.renderEmpty,
      direction = _React$useContext.direction,
      virtual = _React$useContext.virtual,
      dropdownMatchSelectWidth = _React$useContext.dropdownMatchSelectWidth;

  var size = React.useContext(_SizeContext["default"]);
  (0, _devWarning["default"])(multiple !== false || !treeCheckable, 'TreeSelect', '`multiple` will always be `true` when `treeCheckable` is true');
  var prefixCls = getPrefixCls('select', customizePrefixCls);
  var treePrefixCls = getPrefixCls('select-tree', customizePrefixCls);
  var treeSelectPrefixCls = getPrefixCls('tree-select', customizePrefixCls);
  var mergedDropdownClassName = (0, _classnames["default"])(dropdownClassName, "".concat(treeSelectPrefixCls, "-dropdown"), (0, _defineProperty2["default"])({}, "".concat(treeSelectPrefixCls, "-dropdown-rtl"), direction === 'rtl'));
  var isMultiple = !!(treeCheckable || multiple);
  var mergedLabelInValue = treeCheckStrictly || labelInValue; // ===================== internal value ============

  var _useMergedState = (0, _useMergedState3["default"])(defaultValue, {
    value: propsValue
  }),
      _useMergedState2 = (0, _slicedToArray2["default"])(_useMergedState, 2),
      internalValue = _useMergedState2[0],
      setInternalValue = _useMergedState2[1];

  var _React$useState = React.useState(internalValue),
      _React$useState2 = (0, _slicedToArray2["default"])(_React$useState, 2),
      panelValue = _React$useState2[0],
      setPanelValue = _React$useState2[1];

  var _React$useState3 = React.useState(false),
      _React$useState4 = (0, _slicedToArray2["default"])(_React$useState3, 2),
      dropdownOpen = _React$useState4[0],
      setDropdownOpen = _React$useState4[1]; // 是否对外触发 onChange 事件的 lock 锁


  var isConfirmRef = React.useRef(!showConfirm);

  var toggleConfirmRef = function toggleConfirmRef(flag) {
    isConfirmRef.current = flag;
  };

  React.useEffect(function () {
    toggleConfirmRef(!showConfirm);
  }, [showConfirm]);
  React.useEffect(function () {
    setPanelValue(internalValue);
  }, [internalValue]);

  var onInternalChange = function onInternalChange(value, labelList, extra) {
    setPanelValue(value); // 如果没有 showConfirm 或者通过点击多选选项的 x 删除的，则也要触发 onChange
    // 判断是点击多选选项的 x 通过 useRef 来判断标记，同时不会因为使用 setState 的 api 导致重新 render

    if (!showConfirm || isConfirmRef.current) {
      setInternalValue(value);
      onChange === null || onChange === void 0 ? void 0 : onChange(value, labelList, extra);
      toggleConfirmRef(false);
    }
  };

  var setInnerValue = function setInnerValue(_ref) {
    var value = _ref.value,
        labelList = _ref.labelList,
        extra = _ref.extra,
        triggerChange = _ref.triggerChange;
    setPanelValue(value);

    if (triggerChange && labelList && extra) {
      var finalExtra = showConfirm ? (0, _extends2["default"])((0, _extends2["default"])({}, extra), {
        preValue: (0, _utils.toLabeledValues)(internalValue)
      }) : extra;
      setInternalValue(value);
      onChange === null || onChange === void 0 ? void 0 : onChange(value, labelList, finalExtra);
    }
  };

  var resetValue = function resetValue() {
    setPanelValue(internalValue);
  };

  var onInternalDropdownVisibleChange = function onInternalDropdownVisibleChange(open) {
    setDropdownOpen(open); // 有 showConfirm 的情况下，如果没有点击确定，则会还原值；点击确定的情况下，internalValue 也是最新的了，等于没有变化

    if (showConfirm && !open) {
      setPanelValue(internalValue);
    }

    onDropdownVisibleChange === null || onDropdownVisibleChange === void 0 ? void 0 : onDropdownVisibleChange(open);
  }; // ===================== Form =====================


  var _React$useContext2 = React.useContext(_context.FormItemInputContext),
      contextStatus = _React$useContext2.status,
      hasFeedback = _React$useContext2.hasFeedback,
      feedbackIcon = _React$useContext2.feedbackIcon;

  var mergedStatus = (0, _statusUtils.getMergedStatus)(contextStatus, customStatus); // ===================== Icons =====================

  var _getIcons = (0, _iconUtil["default"])((0, _extends2["default"])((0, _extends2["default"])({}, props), {
    multiple: isMultiple,
    showArrow: showArrow,
    hasFeedback: hasFeedback,
    feedbackIcon: feedbackIcon,
    prefixCls: prefixCls,
    dropDownVisible: dropdownOpen
  })),
      suffixIcon = _getIcons.suffixIcon,
      removeIcon = _getIcons.removeIcon,
      clearIcon = _getIcons.clearIcon; // ===================== Empty =====================


  var mergedNotFound;

  if (notFoundContent !== undefined) {
    mergedNotFound = notFoundContent;
  } else {
    mergedNotFound = renderEmpty('Select');
  } // ==================== Render =====================


  var selectProps = (0, _omit["default"])(props, ['suffixIcon', 'itemIcon', 'removeIcon', 'clearIcon', 'switcherIcon']);
  var mergedSize = customizeSize || size;
  var mergedClassName = (0, _classnames["default"])(!customizePrefixCls && treeSelectPrefixCls, (_classNames2 = {}, (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-lg"), mergedSize === 'large'), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-sm"), mergedSize === 'small'), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-rtl"), direction === 'rtl'), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-borderless"), !bordered), (0, _defineProperty2["default"])(_classNames2, "".concat(prefixCls, "-simple"), selectorSimpleMode && isMultiple), _classNames2), (0, _statusUtils.getStatusClassNames)(prefixCls, mergedStatus, hasFeedback), className);
  var rootPrefixCls = getPrefixCls();

  var customerSelector = function customerSelector(_ref2) {
    var getDisplayValues = _ref2.getDisplayValues,
        onOptionSelect = _ref2.onOptionSelect;
    return /*#__PURE__*/React.createElement(_Selector["default"], {
      internalValue: internalValue,
      getDisplayValues: getDisplayValues,
      selectorSimpleMode: selectorSimpleMode,
      multiple: multiple || !!treeCheckable,
      placeholder: placeholder,
      maxTagCount: maxTagCount,
      maxTagPlaceholder: maxTagPlaceholder,
      labelInValue: mergedLabelInValue,
      onOptionSelect: onOptionSelect,
      toggleConfirm: toggleConfirmRef,
      prefixIcon: prefixIcon
    });
  };

  var optionRender = function optionRender(tree, _ref3) {
    var treeData = _ref3.treeData,
        keyEntities = _ref3.keyEntities,
        fieldNames = _ref3.fieldNames,
        toggleOpen = _ref3.toggleOpen,
        onSelect = _ref3.onSelect,
        triggerChange = _ref3.triggerChange;
    return /*#__PURE__*/React.createElement(_WrapperTree["default"], {
      tree: tree,
      prefixCls: prefixCls,
      treePrefixCls: treePrefixCls,
      treeData: treeData,
      keyEntities: keyEntities,
      fieldNames: fieldNames,
      statistics: statistics,
      multiple: multiple || !!treeCheckable,
      noDataDesc: noDataDesc,
      noDataTitle: noDataTitle,
      noResultDesc: noResultDesc,
      noResultTitle: noResultTitle,
      value: panelValue,
      setInnerValue: setInnerValue,
      showSearch: showSearch,
      searchValue: searchValue,
      onSearch: onSearch,
      toggleOpen: toggleOpen,
      treeCheckStrictly: treeCheckStrictly,
      showCheckAll: showCheckAll,
      showConfirm: showConfirm,
      treeNodeFilterProp: treeNodeFilterProp,
      filterTreeNode: filterTreeNode,
      onSelect: onSelect,
      triggerChange: triggerChange,
      dropdownOptLeft: dropdownOptLeft,
      dropdownSearchPlaceholder: dropdownSearchPlaceholder,
      selectAllText: selectAllText,
      cancelText: cancelText,
      okText: okText,
      treeNodeLabelProp: treeNodeLabelProp,
      autoClearSearchValue: autoClearSearchValue,
      resetValue: resetValue,
      filterSort: filterSort
    });
  };

  return /*#__PURE__*/React.createElement(_rcTreeSelect["default"], (0, _extends2["default"])({
    virtual: virtual,
    dropdownMatchSelectWidth: dropdownMatchSelectWidth,
    showArrow: showArrow,
    treeCheckStrictly: treeCheckStrictly,
    labelInValue: labelInValue,
    placeholder: placeholder
  }, selectProps, {
    ref: ref,
    prefixCls: prefixCls,
    className: mergedClassName,
    treeClassName: (0, _classnames["default"])((0, _defineProperty2["default"])({}, "".concat(treePrefixCls, "-select-line"), !multiple && !treeCheckable)),
    listHeight: listHeight,
    listItemHeight: listItemHeight,
    treeCheckable: multiple || treeCheckable ? /*#__PURE__*/React.createElement("span", {
      className: "".concat(prefixCls, "-tree-checkbox-inner")
    }) : treeCheckable,
    treeLine: !!treeLine,
    inputIcon: suffixIcon,
    multiple: multiple,
    removeIcon: removeIcon,
    clearIcon: clearIcon,
    switcherIcon: function switcherIcon(nodeProps) {
      return (0, _iconUtil2["default"])(treePrefixCls, _switcherIcon, treeLine, nodeProps);
    },
    showTreeIcon: treeIcon,
    showSearch: false,
    notFoundContent: mergedNotFound,
    getPopupContainer: getPopupContainer || getContextPopupContainer,
    treeMotion: null,
    dropdownClassName: mergedDropdownClassName,
    choiceTransitionName: (0, _motion.getTransitionName)(rootPrefixCls, '', choiceTransitionName),
    transitionName: (0, _motion.getTransitionName)(rootPrefixCls, 'slide-up', transitionName),
    value: panelValue,
    onChange: onInternalChange,
    optionRender: optionRender // @ts-ignore
    ,
    customerSelector: customerSelector,
    onDropdownVisibleChange: onInternalDropdownVisibleChange,
    getRawInputElement: selectorRender,
    dropdownAutoClose: !showConfirm,
    treeNodeLabelProp: treeNodeLabelProp
  }));
};

var TreeSelectRef = /*#__PURE__*/React.forwardRef(InternalTreeSelect);
var TreeSelect = TreeSelectRef;
TreeSelect.TreeNode = _rcTreeSelect.TreeNode;
TreeSelect.SHOW_ALL = _rcTreeSelect.SHOW_ALL;
TreeSelect.SHOW_PARENT = _rcTreeSelect.SHOW_PARENT;
TreeSelect.SHOW_CHILD = _rcTreeSelect.SHOW_CHILD;
var _default = TreeSelect;
exports["default"] = _default;
