// sensd/lib/upload/UploadList/ItemProgress.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _progress = _interopRequireDefault(require("../../progress"));

var _utils = require("../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var ItemProgress = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _classNames;

  var type = _ref.type,
      prefixCls = _ref.prefixCls,
      locale = _ref.locale,
      file = _ref.file,
      cancelIcon = _ref.cancelIcon,
      errorDom = _ref.errorDom,
      progressProps = _ref.progress,
      className = _ref.className;
  var progressClassName = (0, _classnames["default"])((_classNames = {}, (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-list-item-progress"), true), (0, _defineProperty2["default"])(_classNames, "".concat(prefixCls, "-list-item-progress-type-").concat(type), true), _classNames), className);

  if (type === 'picture') {
    return /*#__PURE__*/React.createElement("div", {
      className: progressClassName,
      ref: ref
    }, 'percent' in file ? /*#__PURE__*/React.createElement(_progress["default"], (0, _extends2["default"])({
      width: 20,
      strokeWidth: 10,
      type: "circle",
      status: file.status === 'error' ? 'exception' : 'normal'
    }, progressProps, {
      percent: file.percent
    })) : null, file.status === 'error' ? errorDom : cancelIcon);
  }

  var fileSizeUnit = (0, _utils.getSuitableFileUnit)(file.size || 0);
  return /*#__PURE__*/React.createElement("div", {
    className: progressClassName,
    ref: ref
  }, file.status === 'uploading' && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-list-item-status")
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("span", null, locale.uploading), file.size && 'percent' in file ? /*#__PURE__*/React.createElement("span", null, " ".concat((0, _utils.formatSizeToSuitableUnit)(file.size * (file.percent || 0) / 100, fileSizeUnit), " "), "/ ", (0, _utils.formatSizeToSuitableUnit)(file.size, fileSizeUnit)) : null), cancelIcon), file.status === 'error' && errorDom, 'percent' in file || file.status === 'error' ? /*#__PURE__*/React.createElement(_progress["default"], (0, _extends2["default"])({
    type: "line",
    strokeWidth: 2,
    status: file.status === 'error' ? 'exception' : 'normal'
  }, progressProps, {
    percent: file.status === 'error' ? 100 : file.percent
  })) : null);
});
var _default = ItemProgress;
exports["default"] = _default;
