// sensd/lib/select/selector/MultipleSelector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _CloseOutlined = _interopRequireDefault(require("@sensd/icons/CloseOutlined"));

var _tag = _interopRequireDefault(require("../../tag"));

var _LocaleReceiver = _interopRequireDefault(require("../../locale-provider/LocaleReceiver"));

var _default2 = _interopRequireDefault(require("../../locale/default"));

var _tooltip = _interopRequireDefault(require("../../tooltip"));

var _typography = _interopRequireDefault(require("../../typography"));

var onPreventMouseDown = function onPreventMouseDown(event) {
  event.preventDefault();
  event.stopPropagation();
}; // @ts-ignore


var MultipleSelector = function MultipleSelector(props) {
  var selectorSimpleMode = props.selectorSimpleMode,
      displayValue = props.displayValue,
      _onClose = props.onClose,
      disabled = props.disabled;

  if (selectorSimpleMode) {
    return /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
      componentName: "Select",
      defaultLocale: _default2["default"].Select
    }, function (locale) {
      return /*#__PURE__*/_react["default"].createElement(_tooltip["default"], {
        title: displayValue === null || displayValue === void 0 ? void 0 : displayValue.map(function (item) {
          return item.label;
        }).join('、')
      }, /*#__PURE__*/_react["default"].createElement("div", null, locale.selectedCount.replace(/\${count}/, displayValue === null || displayValue === void 0 ? void 0 : displayValue.length)));
    });
  }

  if (!displayValue || displayValue.length === 0) return null;
  return displayValue === null || displayValue === void 0 ? void 0 : displayValue.map(function (item) {
    return /*#__PURE__*/_react["default"].createElement(_tag["default"], {
      key: item.value,
      closable: item.closable !== false,
      onClose: function onClose() {
        return _onClose(item.value);
      },
      disabled: disabled || item.disabled,
      closeIcon: /*#__PURE__*/_react["default"].createElement(_CloseOutlined["default"], {
        onMouseDown: onPreventMouseDown
      })
    }, /*#__PURE__*/_react["default"].createElement(_typography["default"].Text, {
      ellipsis: {
        tooltip: item.label
      }
    }, item.label));
  });
};

var _default = MultipleSelector;
exports["default"] = _default;
