// sensd/lib/select-transfer/selectedPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CloseOutlined = _interopRequireDefault(require("@sensd/icons/CloseOutlined"));

var _rcVirtualList = _interopRequireDefault(require("rc-virtual-list"));

var _configProvider = require("../config-provider");

var _LocaleReceiver = _interopRequireDefault(require("../locale-provider/LocaleReceiver"));

var _default2 = _interopRequireDefault(require("../locale/default"));

var _typography = _interopRequireDefault(require("../typography"));

var _result = _interopRequireDefault(require("../result"));

var _button = _interopRequireDefault(require("../button"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var NoData = _result["default"].NoData;

var SelectedPanel = function SelectedPanel(props) {
  var customizePrefixCls = props.prefixCls,
      _props$options = props.options,
      options = _props$options === void 0 ? [] : _props$options,
      contentHeight = props.contentHeight,
      maxCount = props.maxCount,
      _props$clearConfirm = props.clearConfirm,
      clearConfirm = _props$clearConfirm === void 0 ? false : _props$clearConfirm,
      onChange = props.onChange;

  var _useContext = (0, _react.useContext)(_configProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('select-transfer', customizePrefixCls);

  var listRef = _react["default"].useRef(null); // 禁用清空按钮


  var disableClear = options.length === (options === null || options === void 0 ? void 0 : options.filter(function (_ref) {
    var disabled = _ref.disabled;
    return disabled;
  }).length);
  var onDeleteClick = (0, _react.useCallback)(function (delItem) {
    if (delItem.disabled) return;
    var newData = options.filter(function (_ref2) {
      var value = _ref2.value;
      return value !== delItem.value;
    });
    var newVals = newData.map(function (_ref3) {
      var value = _ref3.value;
      return value;
    });
    onChange === null || onChange === void 0 ? void 0 : onChange(newVals, newData);
  }, [onChange, options]);
  var onClearClick = (0, _react.useCallback)(function (e) {
    e.stopPropagation();
    var newData = options.filter(function (_ref4) {
      var disabled = _ref4.disabled;
      return disabled;
    });
    var newVals = newData.map(function (_ref5) {
      var value = _ref5.value;
      return value;
    });
    onChange === null || onChange === void 0 ? void 0 : onChange(newVals, newData);
  }, [onChange]);

  var renderEmpty = function renderEmpty(locale) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-selected-noData")
    }, /*#__PURE__*/_react["default"].createElement(NoData, {
      title: locale.noData,
      type: "component",
      subTitle: /*#__PURE__*/_react["default"].createElement("span", null, locale.emptyText)
    }));
  };

  var renderFooter = function renderFooter(locale) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-selected-footer")
    }, (0, _typeof2["default"])(clearConfirm) === 'object' ? clearConfirm : /*#__PURE__*/_react["default"].createElement(_button["default"], {
      type: "link",
      linkType: "normal",
      onClick: onClearClick,
      disabled: disableClear
    }, locale.clearAll), /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-selected-count")
    }, /*#__PURE__*/_react["default"].createElement("span", null, locale.selectedCount.replace(/\${count}/, String(options.length))), maxCount !== undefined ? /*#__PURE__*/_react["default"].createElement("span", null, locale.maxCount.replace(/\${max}/, String(maxCount))) : null));
  };

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefixCls, "-selected")
  }, (options === null || options === void 0 ? void 0 : options.length) === 0 ? /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
    componentName: "SelectTransfer",
    defaultLocale: _default2["default"].SelectTransfer
  }, renderEmpty) : /*#__PURE__*/_react["default"].createElement("ul", {
    className: "".concat(prefixCls, "-selected-container")
  }, /*#__PURE__*/_react["default"].createElement(_rcVirtualList["default"], {
    ref: listRef,
    itemKey: "label",
    data: options,
    height: contentHeight - 52,
    fullHeight: false
  }, function (item, index) {
    return /*#__PURE__*/_react["default"].createElement("li", {
      key: index,
      className: (0, _classnames["default"])((0, _defineProperty2["default"])({}, "".concat(prefixCls, "-selected-item-disabled"), item === null || item === void 0 ? void 0 : item.disabled), "".concat(prefixCls, "-selected-item"))
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "".concat(prefixCls, "-selected-item-title")
    }, /*#__PURE__*/_react["default"].createElement("p", {
      className: "".concat(prefixCls, "-selected-item-label")
    }, /*#__PURE__*/_react["default"].createElement(_typography["default"].Text, {
      ellipsis: {
        tooltip: true
      },
      className: "".concat(prefixCls, "-ellipsis")
    }, item === null || item === void 0 ? void 0 : item.label)), /*#__PURE__*/_react["default"].createElement(_button["default"], {
      type: "link",
      linkType: "normal",
      icon: /*#__PURE__*/_react["default"].createElement(_CloseOutlined["default"], null),
      disabled: item === null || item === void 0 ? void 0 : item.disabled,
      onClick: function onClick() {
        return onDeleteClick(item);
      }
    })), (item === null || item === void 0 ? void 0 : item.detail) && /*#__PURE__*/_react["default"].createElement("span", {
      className: "".concat(prefixCls, "-selected-item-detail")
    }, /*#__PURE__*/_react["default"].createElement(_typography["default"].Text, {
      ellipsis: {
        tooltip: true
      },
      className: "".concat(prefixCls, "-ellipsis")
    }, item === null || item === void 0 ? void 0 : item.detail)));
  })), /*#__PURE__*/_react["default"].createElement(_LocaleReceiver["default"], {
    componentName: "SelectTransfer",
    defaultLocale: _default2["default"].SelectTransfer
  }, renderFooter));
};

var _default = SelectedPanel;
exports["default"] = _default;
