// sensd/es/select/CustomDropdown.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React, { useState, useEffect, useMemo } from 'react';
import SelectPanel from '../select-panel';
import LocaleReceiver from '../locale-provider/LocaleReceiver';
import defaultLocale from '../locale/default';
import Empty, { EmptyType } from '../select-panel/Empty';
import Button from '../button';

var CustomDropdown = function CustomDropdown(props) {
  var menu = props.menu,
      flattenOptions = props.flattenOptions,
      valueOptions = props.valueOptions,
      mergedOptions = props.options,
      rawValues = props.rawValues,
      multiple = props.multiple,
      noResultDesc = props.noResultDesc,
      noDataTitle = props.noDataTitle,
      noDataDesc = props.noDataDesc,
      noResultTitle = props.noResultTitle,
      prefixCls = props.prefixCls,
      searchValue = props.searchValue,
      onSearch = props.onSearch,
      onOK = props.onOK,
      onCheckAll = props.onCheckAll,
      showCheckAll = props.showCheckAll,
      showConfirm = props.showConfirm,
      onCancel = props.onCancel,
      _props$statistics = props.statistics,
      statistics = _props$statistics === void 0 ? true : _props$statistics,
      showSearch = props.showSearch,
      searchSuffix = props.searchSuffix,
      notFoundContent = props.notFoundContent,
      dropdownSearchPlaceholder = props.dropdownSearchPlaceholder,
      customLeftOpt = props.customLeftOpt,
      selectAllText = props.selectAllText,
      cancelText = props.cancelText,
      okText = props.okText; // 半选

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      indeterminate = _useState2[0],
      setIndeterminate = _useState2[1]; // 全选


  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      checkAll = _useState4[0],
      setCheckAll = _useState4[1];

  var onCheckboxChange = function onCheckboxChange() {
    var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    var disabledSelectedValues = value.filter(function (val) {
      var opt = valueOptions.get(val);
      return opt === null || opt === void 0 ? void 0 : opt.disabled;
    });
    var data = mergedOptions.filter(function (item) {
      return !item.disabled;
    }).map(function (item) {
      return item.value;
    });
    return {
      filteredValues: [].concat(_toConsumableArray(disabledSelectedValues), _toConsumableArray(data)),
      disabledSelectedValues: disabledSelectedValues
    };
  };

  var handleCheckAll = function handleCheckAll(e) {
    var checked = e.target.checked;
    setCheckAll(checked);

    var _onCheckboxChange = onCheckboxChange(_toConsumableArray(rawValues)),
        filteredValues = _onCheckboxChange.filteredValues,
        disabledSelectedValues = _onCheckboxChange.disabledSelectedValues;

    var values = checked ? filteredValues : disabledSelectedValues;
    onCheckAll === null || onCheckAll === void 0 ? void 0 : onCheckAll(values);
  };

  useEffect(function () {
    if (multiple) {
      var size = rawValues.size; // 半选

      setIndeterminate(!!size && size < mergedOptions.length); // 全选

      var _onCheckboxChange2 = onCheckboxChange(_toConsumableArray(rawValues)),
          filteredValues = _onCheckboxChange2.filteredValues;

      setCheckAll(size === filteredValues.length);
    }
  }, [rawValues, mergedOptions]);

  var _useMemo = useMemo(function () {
    return [/*#__PURE__*/React.createElement("div", {
      key: EmptyType.noData,
      className: "".concat(prefixCls, "-nodata")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noData,
      noDataTitle: noDataTitle,
      noDataDesc: noDataDesc
    })), /*#__PURE__*/React.createElement("div", {
      key: EmptyType.noResult,
      className: "".concat(prefixCls, "-no-result")
    }, /*#__PURE__*/React.createElement(Empty, {
      type: EmptyType.noResult,
      noResultTitle: noResultTitle,
      noResultDesc: noResultDesc
    }))];
  }, [noDataDesc]),
      _useMemo2 = _slicedToArray(_useMemo, 2),
      noDataComp = _useMemo2[0],
      noResultComp = _useMemo2[1];

  var count = flattenOptions.length;

  var handleOk = function handleOk() {
    var values = _toConsumableArray(rawValues);

    if (multiple) {
      onOK === null || onOK === void 0 ? void 0 : onOK(values, values.map(function (item) {
        return valueOptions.get(item);
      }));
      return;
    }

    var singleVal = values[0];
    onOK === null || onOK === void 0 ? void 0 : onOK(singleVal, valueOptions.get(singleVal));
  };

  var customRightOpt = /*#__PURE__*/React.createElement(LocaleReceiver, {
    componentName: "Select",
    defaultLocale: defaultLocale.Select
  }, function (locale) {
    return showConfirm && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, {
      style: {
        marginRight: 12
      },
      size: "small",
      onClick: onCancel
    }, cancelText || locale.cancelText), /*#__PURE__*/React.createElement(Button, {
      type: "primary",
      size: "small",
      onClick: handleOk
    }, okText || locale.okText));
  });
  var hasFooter = showCheckAll || showConfirm || customLeftOpt; // 如果是单选，只显示checkbox，则不展示 footer

  if (!multiple && showCheckAll && !showConfirm) {
    hasFooter = false;
  }

  var customContent = notFoundContent !== undefined ? notFoundContent : noDataComp;

  if ((mergedOptions === null || mergedOptions === void 0 ? void 0 : mergedOptions.length) > 0) {
    customContent = undefined;
  }

  return /*#__PURE__*/React.createElement(SelectPanel, {
    query: searchValue,
    count: count,
    onSearch: onSearch,
    indeterminate: indeterminate,
    checkAll: checkAll,
    onCheckAll: handleCheckAll,
    showCheckAll: showCheckAll,
    footer: hasFooter,
    statistics: statistics,
    multiple: multiple,
    customContent: customContent,
    noResult: searchValue && count === 0 && noResultComp,
    customRightOpt: customRightOpt,
    showSearch: showSearch,
    searchSuffix: searchSuffix,
    placeholder: dropdownSearchPlaceholder,
    customLeftOpt: customLeftOpt,
    selectAllText: selectAllText
  }, menu);
};

export default CustomDropdown;
