// @sensd/result-svg/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// This index.ts file is generated automatically.
var LoadingFailComponent_1 = require("./asn/LoadingFailComponent");
Object.defineProperty(exports, "LoadingFailComponent", { enumerable: true, get: function () { return LoadingFailComponent_1.default; } });
var NetworkErrorComponent_1 = require("./asn/NetworkErrorComponent");
Object.defineProperty(exports, "NetworkErrorComponent", { enumerable: true, get: function () { return NetworkErrorComponent_1.default; } });
var NetworkErrorPage_1 = require("./asn/NetworkErrorPage");
Object.defineProperty(exports, "NetworkErrorPage", { enumerable: true, get: function () { return NetworkErrorPage_1.default; } });
var NoDataComponent_1 = require("./asn/NoDataComponent");
Object.defineProperty(exports, "NoDataComponent", { enumerable: true, get: function () { return NoDataComponent_1.default; } });
var NoDataPage_1 = require("./asn/NoDataPage");
Object.defineProperty(exports, "NoDataPage", { enumerable: true, get: function () { return NoDataPage_1.default; } });
var NoFoundPage_1 = require("./asn/NoFoundPage");
Object.defineProperty(exports, "NoFoundPage", { enumerable: true, get: function () { return NoFoundPage_1.default; } });
var NoResultComponent_1 = require("./asn/NoResultComponent");
Object.defineProperty(exports, "NoResultComponent", { enumerable: true, get: function () { return NoResultComponent_1.default; } });
var NoResultPage_1 = require("./asn/NoResultPage");
Object.defineProperty(exports, "NoResultPage", { enumerable: true, get: function () { return NoResultPage_1.default; } });
var UnauthorizedComponent_1 = require("./asn/UnauthorizedComponent");
Object.defineProperty(exports, "UnauthorizedComponent", { enumerable: true, get: function () { return UnauthorizedComponent_1.default; } });
var UnauthorizedPage_1 = require("./asn/UnauthorizedPage");
Object.defineProperty(exports, "UnauthorizedPage", { enumerable: true, get: function () { return UnauthorizedPage_1.default; } });
var UpgradePage_1 = require("./asn/UpgradePage");
Object.defineProperty(exports, "UpgradePage", { enumerable: true, get: function () { return UpgradePage_1.default; } });
